<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Tidy Transcriptomics • rpharma2021tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Tidy Transcriptomics">
<meta property="og:description" content="rpharma2021tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rpharma2021tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/background.html">Background</a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tidytranscriptomics-workshops/rpharma2021_tidytranscriptomics">
    <span class="fab fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidytranscriptomics_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Tidy Transcriptomics</h1>
                        <h4 class="author">Maria Doyle, Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Stefano Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/tidytranscriptomics-workshops/rpharma2021_tidytranscriptomics/blob/master/vignettes/tidytranscriptomics.Rmd"><code>vignettes/tidytranscriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>tidytranscriptomics.Rmd</code></div>

    </div>

    
    
<style type="text/css">
# Formatting for challenges
.challenge {
background-color: #fff4d4;
}
.challenge code {
    background-color: #fff4d4;
}
</style>
<div id="workshop-introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-introduction" class="anchor"></a>Workshop introduction</h1>
<p><img src="../inst/vignettes/tidybulk_logo.png" width="100px"></p>
<div id="instructors" class="section level2">
<h2 class="hasAnchor">
<a href="#instructors" class="anchor"></a>Instructors</h2>
<p>Dr. Stefano Mangiola is currently a Postdoctoral researcher in the laboratory of Prof. Tony Papenfuss. His background spans from biotechnology to bioinformatics and biostatistics. His research focuses on prostate and breast tumour microenvironment, the development of statistical model for the analysis of RNA sequencing data, and data analysis and visualisation interfaces.</p>
<p>Dr. Maria Doyle is the Application and Training Specialist for Research Computing at the Peter MacCallum Cancer Centre in Melbourne, Australia. She has a PhD in Molecular Biology and currently works in bioinformatics and data science education and training. She is passionate about supporting researchers, reproducible research, open source and tidy data.</p>
</div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>This tutorial will present how to perform analysis of single-cell and bulk RNA sequencing data following the tidy data paradigm. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions.</p>
<p>This can be achieved with the integration of packages present in the R CRAN and Bioconductor ecosystem, including <a href="https://stemangiola.github.io/tidyseurat/">tidyseurat</a>, <a href="https://stemangiola.github.io/tidySingleCellExperiment/">tidySingleCellExperiment</a>, <a href="https://stemangiola.github.io/tidybulk/">tidybulk</a>, <a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> and <a href="https://www.tidyverse.org/">tidyverse</a>. These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data representation and analysis.</p>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li>Familiarity with tidyverse syntax</li>
<li>Some familiarity with bulk RNA-seq and single cell RNA-seq</li>
</ul>
<p>Strongly recommended background reading:</p>
<p><a href="https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html" class="uri">https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html</a> <a href="https://towardsdatascience.com/coding-in-r-nest-and-map-your-way-to-efficient-code-4e44ba58ee4a" class="uri">https://towardsdatascience.com/coding-in-r-nest-and-map-your-way-to-efficient-code-4e44ba58ee4a</a> by Rebecca O’Dwyer <a href="https://finnstats.com/index.php/2021/04/02/tidyverse-in-r/" class="uri">https://finnstats.com/index.php/2021/04/02/tidyverse-in-r/</a></p>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<p>The workshop format is a 3 hour session consisting of hands-on demos, exercises and Q&amp;A.</p>
<p>Guide</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Part 1 Bulk RNA-seq Core</strong></td>
<td></td>
</tr>
<tr class="even">
<td><em>Hands-on Demos + Exercises</em></td>
<td>90m</td>
</tr>
<tr class="odd">
<td>    Differential gene expression</td>
<td></td>
</tr>
<tr class="even">
<td>    Cell type composition analysis</td>
<td></td>
</tr>
<tr class="odd">
<td><strong>Part 2 Single-cell RNA-seq</strong></td>
<td></td>
</tr>
<tr class="even">
<td><em>Hands-on Demos + Exercises</em></td>
<td>90m</td>
</tr>
<tr class="odd">
<td>    Single-cell analysis</td>
<td></td>
</tr>
<tr class="even">
<td>    Pseudobulk analysis</td>
<td></td>
</tr>
<tr class="odd">
<td>Total</td>
<td>180m</td>
</tr>
</tbody>
</table>
<p><br><strong>Format</strong>: Hands on demos, exercises plus Q&amp;A <strong>Interact</strong>: Zoom chat <br><br></p>
</div>
<div id="workshop-goals-and-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h3>
<p>In exploring and analysing RNA sequencing data, there are a number of key concepts, such as filtering, scaling, dimensionality reduction, hypothesis testing, clustering and visualisation, that need to be understood. These concepts can be intuitively explained to new users, however, (i) the use of a heterogeneous vocabulary and jargon by methodologies/algorithms/packages, (ii) the complexity of data wrangling, and (iii) the coding burden, impede effective learning of the statistics and biology underlying an informed RNA sequencing analysis.</p>
<p>The tidytranscriptomics approach to RNA sequencing data analysis abstracts out the coding-related complexity and provides tools that use an intuitive and jargon-free vocabulary, enabling focus on the statistical and biological challenges.</p>
<div id="learning-goals" class="section level4">
<h4 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h4>
<ul>
<li>To understand the key concepts and steps of RNA sequencing data analysis</li>
<li>To approach data representation and analysis though a tidy data paradigm, integrating tidyverse with tidybulk, tidyseurat, tidySingleCellExperiment and tidyHeatmap.</li>
</ul>
</div>
<div id="learning-objectives" class="section level4">
<h4 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h4>
<ul>
<li>Recall the key concepts of RNA sequencing data analysis</li>
<li>Apply the concepts to publicly available data</li>
<li>Create plots that summarise the information content of the data and analysis results</li>
</ul>
</div>
</div>
</div>
<div id="what-is-transcriptomics" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-transcriptomics" class="anchor"></a>What is transcriptomics?</h2>
<p>“The transcriptome is the set of all RNA transcripts, including coding and non-coding, in an individual or a population of cells”</p>
<p><em>Wikipedia</em></p>
<p><img src="../inst/vignettes/transcriptomics.jpg" width="600px"></p>
</div>
<div id="why-use-transcriptomics" class="section level2">
<h2 class="hasAnchor">
<a href="#why-use-transcriptomics" class="anchor"></a>Why use transcriptomics?</h2>
<ul>
<li>Genome (DNA) pretty stable</li>
<li>Proteome (proteins) harder to measure</li>
<li>Transcriptome (RNA) can measure changes in expression of thousands of coding and non-coding transcripts <br><br>
</li>
</ul>
</div>
<div id="possible-experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#possible-experimental-design" class="anchor"></a>Possible experimental design</h2>
<p><img src="../inst/vignettes/ScreenShot2.png" width="600px"></p>
</div>
<div id="how-does-transcriptomics-work" class="section level2">
<h2 class="hasAnchor">
<a href="#how-does-transcriptomics-work" class="anchor"></a>How does transcriptomics work?</h2>
<p><img src="../inst/vignettes/ScreenShot3.png" width="600px"></p>
</div>
<div id="types-of-transcriptomic-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#types-of-transcriptomic-analyses" class="anchor"></a>Types of transcriptomic analyses</h2>
<ul>
<li>
<strong>Differential expression</strong><br>
</li>
<li>
<strong>Cell type composition</strong><br>
</li>
<li>Alternative splicing<br>
</li>
<li>Novel transcript discovery<br>
</li>
<li>Fusions identification<br>
</li>
<li>Variant analysis<br><br> Topics in bold we will see in this workshop <br><br>
</li>
</ul>
</div>
<div id="bulk-rna-sequencing-differential-expression-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#bulk-rna-sequencing-differential-expression-workflow" class="anchor"></a>Bulk RNA sequencing differential expression workflow</h2>
<p><img src="../inst/vignettes/bulk_RNAseq_pipeline.png" width="600px"></p>
</div>
<div id="differences-between-bulk-and-single-cell-rna-sequencing" class="section level2">
<h2 class="hasAnchor">
<a href="#differences-between-bulk-and-single-cell-rna-sequencing" class="anchor"></a>Differences between bulk and single-cell RNA sequencing</h2>
<p><img src="../inst/vignettes/bulk_vs_single.jpg" width="600px"></p>
<p><em>Shalek and Benson, 2017</em></p>
</div>
<div id="single-cell-rna-sequencing-analysis-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#single-cell-rna-sequencing-analysis-workflow" class="anchor"></a>Single-cell RNA sequencing analysis workflow</h2>
<p><img src="../inst/vignettes/single_cell_RNAseq_pipeline.png" width="600px"></p>
</div>
<div id="tidy-data-and-the-tidyverse" class="section level2">
<h2 class="hasAnchor">
<a href="#tidy-data-and-the-tidyverse" class="anchor"></a>Tidy data and the tidyverse</h2>
<p>This workshop demonstrates how to perform analysis of RNA sequencing data following the tidy data paradigm <span class="citation">(Wickham et al. 2014)</span>. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions. For more information, see the R for Data Science chapter on tidy data <a href="https://r4ds.had.co.nz/tidy-data.html">here</a>.</p>
<div class="figure">
<img src="../inst/vignettes/tidydata_1.jpg" alt="[Tidy data for efficiency, reproducibilty and collaboration](https://www.openscapes.org/blog/2020/10/12/tidy-data/)" width="500px"><p class="caption">
<a href="https://www.openscapes.org/blog/2020/10/12/tidy-data/">Tidy data for efficiency, reproducibilty and collaboration</a>
</p>
</div>
<p>The tidyverse is a collection of packages that can be used to tidy, manipulate and visualise data. We’ll use many functions from the tidyverse in this workshop, such as <code>filter</code>, <code>select</code>, <code>mutate</code>, <code>pivot_longer</code> and <code>ggplot</code>.</p>
<div class="figure">
<img src="../inst/vignettes/tidyverse.png" alt="Adapted from [Getting Started with tidyverse in R](https://www.storybench.org/getting-started-with-tidyverse-in-r/)" width="500px"><p class="caption">
Adapted from <a href="https://www.storybench.org/getting-started-with-tidyverse-in-r/">Getting Started with tidyverse in R</a>
</p>
</div>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<div id="cloud" class="section level3">
<h3 class="hasAnchor">
<a href="#cloud" class="anchor"></a>Cloud</h3>
<p>Easiest way to run this material. Only available during workshop.</p>
</div>
<div id="local" class="section level3">
<h3 class="hasAnchor">
<a href="#local" class="anchor"></a>Local</h3>
<p>We will use RStudio Cloud during the RPharma workshop. After the workshop, if you want to install on your own computer, see instructions <a href="https://tidytranscriptomics-workshops.github.io/rpharma2021_tidytranscriptomics/index.html#workshop-package-installation">here</a>.</p>
<p>Alternatively, you can view the material at the workshop webpage <a href="https://tidytranscriptomics-workshops.github.io/rpharma2021_tidytranscriptomics/articles/tidytranscriptomics.html">here</a>.</p>
</div>
</div>
<div id="what-is-tidytranscriptomics" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-tidytranscriptomics" class="anchor"></a>What is tidytranscriptomics?</h2>
<p><a href="https://stemangiola.github.io/tidybulk/">tidybulk</a>, <a href="https://stemangiola.github.io/tidySummarizedExperiment/">tidySummarizedExperiment</a> and <a href="https://stemangiola.github.io/tidySingleCellExperiment/">tidySingleCellExperiment</a> are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data representation and analysis. The roadmap for the development of the tidytranscriptomics packages is shown in the figure below.</p>
<p><img src="../inst/vignettes/roadmap.png" width="600px"></p>
</div>
</div>
<div id="part-1-bulk-rna-sequencing-with-tidysummarizedexperiment-and-tidybulk" class="section level1">
<h1 class="hasAnchor">
<a href="#part-1-bulk-rna-sequencing-with-tidysummarizedexperiment-and-tidybulk" class="anchor"></a>Part 1 Bulk RNA sequencing with tidySummarizedExperiment and tidybulk</h1>
<p><img src="../inst/vignettes/tidybulk_logo.png" width="100px"></p>
<div id="acknowledgements" class="section level3">
<h3 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h3>
<p>Some of the material in Part 1 was adapted from an R RNA sequencing workshop first run <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>. The use of the airway dataset was inspired by the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>.</p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next-generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA sequencing to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA sequencing dataset. Sequenced reads are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today, we will be starting from the count data and showing how differential expression analysis can be performed in a friendly way using the Bioconductor package, tidybulk.</p>
<p>First, let’s load all the packages we will need to analyse the data.</p>
<p><em>Note: you should load the tidybulk</em> and <em>tidySummarizedExperiment</em> libraries after the tidyverse core packages for best integration.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dataset</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>

<span class="co"># tidyverse-friendly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/">GGally</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">tidyHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidybulk">tidybulk</a></span><span class="op">)</span>
<span class="co"># library(tidySummarizedExperiment) # we'll load this below to show what it can do</span></code></pre></div>
<p>Plot settings. Set the colours and theme we will use for our plots.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use colourblind-friendly colours</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Set theme</span>
<span class="va">custom_theme</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
        panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,
        panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
      <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<div id="tidysummarizedexperiment" class="section level3">
<h3 class="hasAnchor">
<a href="#tidysummarizedexperiment" class="anchor"></a>tidySummarizedExperiment</h3>
<p>Here we will perform our analysis using the data from the <em>airway</em> package. The airway data comes from the paper by <span class="citation">(Himes et al. 2014)</span>; and it includes 8 samples from human airway smooth muscle cells, from 4 cell lines. For each cell line treated (with dexamethasone) and untreated (negative control) a sample has undergone RNA sequencing and gene counts have been generated.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load airway RNA sequencing data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">airway</span>
<span class="co">#&gt; class: RangedSummarizedExperiment </span>
<span class="co">#&gt; dim: 64102 8 </span>
<span class="co">#&gt; metadata(1): ''</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span>
<span class="co">#&gt; colData names(9): SampleName cell ... Sample BioSample</span></code></pre></div>
<p>The data in the airway package is a Bioconductor <em>SummarizedExperiment</em> object. For more information see <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">here</a>.</p>
<p>The <code>tidySummarizedExperiment</code> package enables any <em>SummarizedExperiment</em> object to be displayed and manipulated according to tidy data principles, without affecting any <em>SummarizedExperiment</em> behaviour.</p>
<p>If we load the <em>tidySummarizedExperiment</em> package and then view the airway data it now displays as a tibble. A tibble is the tidyverse table format.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load tidySummarizedExperiment package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySummarizedExperiment</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">airway</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 512,816 × 13</span>
<span class="co">#&gt; [90m# Transcripts=64102 | Samples=8 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 3 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   coordinate &lt;list&gt;</span></code></pre></div>
<p>Now we can more easily see the data. The airway object contains information about genes and samples, the first column has the Ensembl gene identifier, the second column has the sample identifier and the third column has the gene transcription abundance. The abundance is the number of reads aligning to the gene in each experimental sample. The remaining columns include sample information. The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from.</p>
<p>We can still interact with the tidy SummarizedExperiment object using commands for SummarizedExperiment objects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assays</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="co">#&gt; List of length 1</span>
<span class="co">#&gt; names(1): counts</span></code></pre></div>
</div>
<div id="tidyverse-commands" class="section level3">
<h3 class="hasAnchor">
<a href="#tidyverse-commands" class="anchor"></a>Tidyverse commands</h3>
<p>And now we can also use tidyverse commands, such as <code>filter</code>, <code>select</code>, <code>group_by</code>, <code>summarise</code> and <code>mutate</code> to explore the tidy SummarizedExperiment object. Some examples are shown below and more can be seen at the tidySummarizedExperiment website <a href="https://stemangiola.github.io/tidySummarizedExperiment/articles/introduction.html#tidyverse-commands-1">here</a>. We can also use the tidyverse pipe <code><a href="https://rdrr.io/pkg/tidyseurat/man/pipe.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps. It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<p>We can use <code>filter</code> to choose rows, for example, to see just the rows for the treated samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">dex</span> <span class="op">==</span> <span class="st">"trt"</span><span class="op">)</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 256,408 × 13</span>
<span class="co">#&gt; [90m# Transcripts=64102 | Samples=4 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    448 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    515 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    211 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     55 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3679 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1062 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    380 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    236 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt; # … with 40 more rows, and 3 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   coordinate &lt;list&gt;</span></code></pre></div>
<p>We can use <code>select</code> to choose columns, for example, to see the sample, cell line and treatment columns.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">cell</span>, <span class="va">dex</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 512,816 × 3</span>
<span class="co">#&gt;    sample     cell   dex  </span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;fct&gt;  &lt;fct&gt;</span>
<span class="co">#&gt;  1 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  2 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  3 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  4 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  5 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  6 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  7 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  8 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  9 SRR1039508 N61311 untrt</span>
<span class="co">#&gt; 10 SRR1039508 N61311 untrt</span>
<span class="co">#&gt; # … with 512,806 more rows</span></code></pre></div>
<p>We can combine <code>group_by</code> and <code>summarise</code> to calculate the total counts for each sample.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">summarise</a></span><span class="op">(</span>total_counts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 8 × 2</span>
<span class="co">#&gt;   sample     total_counts</span>
<span class="co">#&gt;   &lt;chr&gt;             &lt;int&gt;</span>
<span class="co">#&gt; 1 SRR1039508     20637971</span>
<span class="co">#&gt; 2 SRR1039509     18809481</span>
<span class="co">#&gt; 3 SRR1039512     25348649</span>
<span class="co">#&gt; 4 SRR1039513     15163415</span>
<span class="co">#&gt; 5 SRR1039516     24448408</span>
<span class="co">#&gt; 6 SRR1039517     30818215</span>
<span class="co">#&gt; 7 SRR1039520     19126151</span>
<span class="co">#&gt; 8 SRR1039521     21164133</span></code></pre></div>
<p>We can use <code>mutate</code> to create a column. For example, we could create a new sample_name column that contains shorter sample names. We can remove the SRR1039 prefix that’s present in all of the samples, as shorter names can fit better in some of the plots we will create. We can use <code>mutate</code> together with <code>str_replace</code> to remove the SRR1039 string from the sample column.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>sample_name<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># select columns to view    </span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">sample_name</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 512,816 × 2</span>
<span class="co">#&gt;    sample     sample_name</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;      </span>
<span class="co">#&gt;  1 SRR1039508 508        </span>
<span class="co">#&gt;  2 SRR1039508 508        </span>
<span class="co">#&gt;  3 SRR1039508 508        </span>
<span class="co">#&gt;  4 SRR1039508 508        </span>
<span class="co">#&gt;  5 SRR1039508 508        </span>
<span class="co">#&gt;  6 SRR1039508 508        </span>
<span class="co">#&gt;  7 SRR1039508 508        </span>
<span class="co">#&gt;  8 SRR1039508 508        </span>
<span class="co">#&gt;  9 SRR1039508 508        </span>
<span class="co">#&gt; 10 SRR1039508 508        </span>
<span class="co">#&gt; # … with 512,806 more rows</span></code></pre></div>
</div>
</div>
<div id="setting-up-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-data" class="anchor"></a>Setting up the data</h2>
<p>We’ll set up the airway data for our RNA sequencing analysis. We’ll create a column with shorter sample names and a column with gene symbols. We can get the gene symbols for these Ensembl gene ids using the Bioconductor annotation package for human, <code>org.Hs.eg.db</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># setup data workflow</span>
<span class="va">counts</span> <span class="op">&lt;-</span>
  <span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span><span class="fu">org.Hs.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Hs.eg.db/man/org.Hs.egBASE.html">org.Hs.eg.db</a></span>,
    keys <span class="op">=</span> <span class="va">feature</span>,
    keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span>,
    column <span class="op">=</span> <span class="st">"SYMBOL"</span>,
    multiVals <span class="op">=</span> <span class="st">"first"</span>
  <span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span>
<span class="co"># take a look</span>
<span class="va">counts</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 512,816 × 15</span>
<span class="co">#&gt; [90m# Transcripts=64102 | Samples=8 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 5 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   sample_name &lt;chr&gt;, symbol &lt;chr&gt;, coordinate &lt;list&gt;</span></code></pre></div>
</div>
<div id="filtering-lowly-transcribed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-lowly-transcribed-genes" class="anchor"></a>Filtering lowly transcribed genes</h2>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they can interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing the power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>We can perform the filtering using tidybulk <code>keep_abundant</code> or <code>identify_abundant</code>. These functions can use the <em>edgeR</em> <code>filterByExpr</code> function described in <span class="citation">(Law et al. 2016)</span> to automatically identify the genes with adequate abundance for differential expression testing. By default, this will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset, the smallest group size is four (as we have four dex-treated samples versus four untreated). Alternatively, we could use <code>identify_abundant</code> to identify which genes are abundant or not (TRUE/FALSE), rather than just keeping the abundant ones.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filtering counts</span>
<span class="va">counts_filtered</span> <span class="op">&lt;-</span> <span class="va">counts</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_filtered</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 × 16</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 6 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   sample_name &lt;chr&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, coordinate &lt;list&gt;</span></code></pre></div>
<p>After running <code>keep_abundant</code> we have a column called <code>.abundant</code> containing TRUE (<code>identify_abundant</code> would have TRUE/FALSE).</p>
</div>
<div id="scaling-counts-to-normalise" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-counts-to-normalise" class="anchor"></a>Scaling counts to normalise</h2>
<p>Scaling of counts (normalisation) is performed to eliminate uninteresting differences between samples due to sequencing depth or composition. A more detailed explanation can be found <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">here</a>. In the tidybulk package, the function <code>scale_abundance</code> generates scaled counts, with scaling factors calculated on abundant (filtered) transcripts and applied to all transcripts. We can choose from different normalisation methods. Here we will use the default edgeR’s trimmed mean of M values (TMM), <span class="citation">(Robinson and Oshlack 2010)</span>. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scaling counts</span>
<span class="va">counts_scaled</span> <span class="op">&lt;-</span> <span class="va">counts_filtered</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_scaled</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 × 19</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts, counts_scaled[39m</span>
<span class="co">#&gt;    feature sample counts counts_scaled SampleName cell  dex   albut Run  </span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt;         &lt;dbl&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679         961.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467         661.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260         368.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60          84.9 GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251        4601.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433        2028.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519         734.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394         558.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172         243.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112        2989.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; # … with 40 more rows, and 10 more variables: avgLength &lt;int&gt;,</span>
<span class="co">#&gt; #   Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;, sample_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;,</span>
<span class="co">#&gt; #   coordinate &lt;list&gt;</span></code></pre></div>
<p>After we run, we should see some columns have been added at the end. The <code>counts_scaled</code> column contains the scaled counts.</p>
<p>We can visualise the difference in abundance densities before and after scaling. As tidybulk output is compatible with tidyverse, we can simply pipe it into standard tidyverse functions such as <code>filter</code>, <code>pivot_longer</code> and <code>ggplot</code>. We can also take advantage of ggplot’s <code>facet_wrap</code> to easily create multiple plots.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshaping</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample_name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-26-1.png" width="70%"></p>
<p>In this dataset, the distributions of the counts are not very different to each other before scaling, but scaling does make the distributions more similar. If we saw a sample with a very different distribution, we may need to investigate it.</p>
<p>As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make plots easier to generate. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from <span class="citation">(Law et al. 2016)</span>).</p>
<p><strong>tidybulk</strong></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tidybulk</span>
<span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="co"># Prepare data set</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">nsamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>

<span class="co"># Setup graphics</span>
<span class="va">col</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="va">nsamples</span>, <span class="st">"Paired"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Plot raw counts</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.26</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"Counts"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col <span class="op">=</span> <span class="va">col</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span>

<span class="co"># Plot scaled counts</span>
<span class="va">dgList_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">lcpm_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">dgList_norm</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.26</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Counts scaled"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="va">dgList_norm</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col <span class="op">=</span> <span class="va">col</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></code></pre></div>
</div>
<div id="exploratory-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#exploratory-analyses" class="anchor"></a>Exploratory analyses</h2>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h3>
<p>By far, one of the most important plots we make when we analyse RNA sequencing data are principal-component analysis (PCA) or multi-dimensional scaling (MDS) plots. We reduce the dimensions of the data to identify the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code>reduce_dimensions</code> function to calculate the dimensions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get principal components</span>
<span class="va">counts_scal_PCA</span> <span class="op">&lt;-</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 × 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.353     1</span>
<span class="co">#&gt; 2                  0.312     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></code></pre></div>
<pre class="challenge_1 challenge"><code>Challenge: What fraction of variance is explained by PC3?</code></pre>
<p>This joins the result to the counts’ object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Take a look</span>
<span class="va">counts_scal_PCA</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 × 21</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts, counts_scaled[39m</span>
<span class="co">#&gt;    feature sample counts counts_scaled SampleName cell  dex   albut Run  </span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt;         &lt;dbl&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679         961.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467         661.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260         368.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60          84.9 GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251        4601.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433        2028.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519         734.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394         558.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172         243.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112        2989.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; # … with 40 more rows, and 12 more variables: avgLength &lt;int&gt;,</span>
<span class="co">#&gt; #   Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;, sample_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, symbol &lt;chr&gt;,</span>
<span class="co">#&gt; #   .abundant &lt;lgl&gt;, coordinate &lt;list&gt;</span></code></pre></div>
<p>For plotting, we can select just the sample-wise information with <code>pivot_sample</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># take a look</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 8 × 15</span>
<span class="co">#&gt;   sample     SampleName cell    dex   albut Run     avgLength Experiment Sample </span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;fct&gt;      &lt;fct&gt;   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;       &lt;int&gt; &lt;fct&gt;      &lt;fct&gt;  </span>
<span class="co">#&gt; 1 SRR1039508 GSM1275862 N61311  untrt untrt SRR103…       126 SRX384345  SRS508…</span>
<span class="co">#&gt; 2 SRR1039509 GSM1275863 N61311  trt   untrt SRR103…       126 SRX384346  SRS508…</span>
<span class="co">#&gt; 3 SRR1039512 GSM1275866 N052611 untrt untrt SRR103…       126 SRX384349  SRS508…</span>
<span class="co">#&gt; 4 SRR1039513 GSM1275867 N052611 trt   untrt SRR103…        87 SRX384350  SRS508…</span>
<span class="co">#&gt; 5 SRR1039516 GSM1275870 N080611 untrt untrt SRR103…       120 SRX384353  SRS508…</span>
<span class="co">#&gt; 6 SRR1039517 GSM1275871 N080611 trt   untrt SRR103…       126 SRX384354  SRS508…</span>
<span class="co">#&gt; 7 SRR1039520 GSM1275874 N061011 untrt untrt SRR103…       101 SRX384357  SRS508…</span>
<span class="co">#&gt; 8 SRR1039521 GSM1275875 N061011 trt   untrt SRR103…        98 SRX384358  SRS508…</span>
<span class="co">#&gt; # … with 6 more variables: BioSample &lt;fct&gt;, sample_name &lt;chr&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></code></pre></div>
<p>We can now plot the reduced dimensions.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># PCA plot</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, colour <span class="op">=</span> <span class="va">dex</span>, shape <span class="op">=</span> <span class="va">cell</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">sample_name</span><span class="op">)</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-33-1.png" width="70%"></p>
<p>The samples group by treatment on PC1 which is what we hope to see. PC2 separates the N080611 cell line from the other samples, indicating a greater difference between that cell line and the others.</p>
</div>
<div id="hierarchical-clustering-with-heatmaps" class="section level3">
<h3 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmaps" class="anchor"></a>Hierarchical clustering with heatmaps</h3>
<p>An alternative to principal component analysis for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use, <code>keep_variable</code>, to extract the most variable genes which we can then plot with tidyHeatmap.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>

  <span class="co"># extract 500 most variable genes</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span>.abundance <span class="op">=</span> <span class="va">counts_scaled</span>, top <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tibble-methods.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># create heatmap</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span><span class="op">(</span>
    .column <span class="op">=</span> <span class="va">sample_name</span>,
    .row <span class="op">=</span> <span class="va">feature</span>,
    .value <span class="op">=</span> <span class="va">counts_scaled</span>,
    transform <span class="op">=</span> <span class="va">log1p</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; tidyHeatmap says: (once per session) from release 1.2.3 the grouping labels have white background by default. To add color for one-ay grouping specify palette_grouping = list(c("red", "blue"))</span>
<span class="co">#&gt; Warning in add_annotation(., !!.column, type = "tile", palette_discrete</span>
<span class="co">#&gt; = .data@data %&gt;% : tidyHeatmap says: nested/list column are present in your data</span>
<span class="co">#&gt; frame and have been dropped as their unicity cannot be identified by dplyr.</span>

<span class="co">#&gt; Warning in add_annotation(., !!.column, type = "tile", palette_discrete</span>
<span class="co">#&gt; = .data@data %&gt;% : tidyHeatmap says: nested/list column are present in your data</span>
<span class="co">#&gt; frame and have been dropped as their unicity cannot be identified by dplyr.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-34-1.png" width="70%"></p>
<p>In the heatmap, we can see the samples cluster into two groups, treated and untreated, for three of the cell lines, and the cell line (N080611) again is further away from the others.</p>
<p>Tidybulk enables a simplified way of generating a clustered heatmap of variable genes. Compare the code below for tidybulk versus a base R method.</p>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">dgList</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span>
<span class="va">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">var_genes</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>
<span class="va">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="va">logcounts</span><span class="op">[</span><span class="va">select_var</span>, <span class="op">]</span>
<span class="va">colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#fefada"</span><span class="op">)</span>
<span class="va">col.group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"grey"</span><span class="op">)</span><span class="op">[</span><span class="va">group</span><span class="op">]</span>
<span class="fu">gplots</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span><span class="op">(</span><span class="va">highly_variable_lcpm</span>, col <span class="op">=</span> <span class="va">colours</span>, trace <span class="op">=</span> <span class="st">"none"</span>, ColSideColors <span class="op">=</span> <span class="va">col.group</span>, scale <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h2>
<p><em>tidybulk</em> integrates several popular methods for differential transcript abundance testing: the edgeR quasi-likelihood <span class="citation">(Chen, Lun, and Smyth 2016)</span> (tidybulk default method), edgeR likelihood ratio <span class="citation">(McCarthy, Chen, and Smyth 2012)</span>, limma-voom <span class="citation">(Law et al. 2014)</span> and DESeq2 <span class="citation">(Love, Huber, and Anders 2014)</span>. A common question researchers have is which method to choose. With tidybulk, we can easily run multiple methods and compare.</p>
<p>We give <code>test_differential_abundance</code> our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line, and there were no additional factors contributing variance, such as batch differences, we could use the formula <code>~ dex</code>. However, each treated and untreated sample is from a different cell line, so we add the cell line as an additional factor <code>~ dex + cell</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">&lt;-</span>

  <span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>

  <span class="co"># edgeR QLT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerQLT_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># edgeR LRT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"edgeR_likelihood_ratio"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerLR_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># limma-voom</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"limma_voom"</span>,
    prefix <span class="op">=</span> <span class="st">"voom_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># DESeq2</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"deseq2"</span>,
    prefix <span class="op">=</span> <span class="st">"deseq2_"</span>
  <span class="op">)</span>

<span class="co"># take a look</span>

<span class="va">de_all</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 × 43</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts, counts_scaled[39m</span>
<span class="co">#&gt;    feature sample counts counts_scaled SampleName cell  dex   albut Run  </span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt;         &lt;dbl&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679         961.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467         661.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260         368.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60          84.9 GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251        4601.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433        2028.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519         734.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394         558.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172         243.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112        2989.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; # … with 40 more rows, and 34 more variables: avgLength &lt;int&gt;,</span>
<span class="co">#&gt; #   Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;, sample_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, symbol &lt;chr&gt;,</span>
<span class="co">#&gt; #   .abundant &lt;lgl&gt;, edgerQLT_logFC &lt;dbl&gt;, edgerQLT_logCPM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerQLT_F &lt;dbl&gt;, edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_logFC &lt;dbl&gt;, edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_PValue &lt;dbl&gt;, edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;, voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_B &lt;dbl&gt;, deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_lfcSE &lt;dbl&gt;, deseq2_stat &lt;dbl&gt;, deseq2_pvalue &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_padj &lt;dbl&gt;, coordinate &lt;list&gt;</span></code></pre></div>
<p>This outputs the columns from each method such as log-fold change (logFC), false-discovery rate (FDR) and probability value (p-value). logFC is log2(treated/untreated).</p>
<div id="comparison-of-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#comparison-of-methods" class="anchor"></a>Comparison of methods</h3>
<p>We can visually compare the significance for all methods. We will notice that there is some difference between the methods.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_PValue</span>, <span class="va">edgerLR_PValue</span>, <span class="va">voom_P.Value</span>, <span class="va">deseq2_pvalue</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<pre class="challenge_2 challenge"><code>Challenge: Which method detects the most differentially abundant transcripts, p value adjusted for multiple testing &lt;  0.05 (FDR, adj.P.Val, padj)?</code></pre>
</div>
<div id="single-method" class="section level3">
<h3 class="hasAnchor">
<a href="#single-method" class="anchor"></a>Single method</h3>
<p>If we just wanted to run one differential testing method we could do that. The default method is edgeR quasi-likelihood.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_de</span> <span class="op">&lt;-</span> <span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), dexuntrt, cellN061011, cellN080611, cellN61311"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR_quasi_likelihood`</span></code></pre></div>
<p>Tidybulk enables a simplified way of performing an RNA sequencing differential expression analysis (with the benefit of smoothly integrating with ggplot2 and other tidyverse packages). Compare the code for a tidybulk edgeR analysis versus standard edgeR below.</p>
<p><strong>standard edgeR</strong></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="volcano-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h2>
<p>Volcano plots are a useful genome-wide tool for checking that the analysis looks good. Volcano plots enable us to visualise the significance of change (p-value) versus the fold change (logFC). Highly significant genes are towards the top of the plot. We can also colour significant genes, e.g. genes with false-discovery rate &lt; 0.05. To decide which genes are differentially expressed, we usually take a cut-off of 0.05 on the FDR (or adjusted P-value), NOT the raw p-value. This is because we are testing many genes (multiple testing), and the chances of finding differentially expressed genes are very high when you do that many tests. Hence we need to control the false discovery rate, the adjusted p-value column in the results table. That is, if 100 genes are significant at a 5% false discovery rate, we are willing to accept that 5 will be false positives.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># volcano plot, minimal</span>
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, colour <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-41-1.png" width="70%"></p>
<p>We’ll extract the symbols for a few top genes (by P value) to use in a more informative volcano plot, integrating some of the packages in tidyverse.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">topgenes_symbols</span> <span class="op">&lt;-</span>
  <span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>

<span class="va">topgenes_symbols</span>
<span class="co">#&gt; [1] "CACNB2"  "ZBTB16"  "PRSS35"  "PDPN"    "SPARCL1" "DUSP1"</span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Subset data</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>, <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">significant</span>, size <span class="op">=</span> <span class="va">significant</span>, alpha <span class="op">=</span> <span class="va">significant</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Custom scales</span>
  <span class="va">custom_theme</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">#&gt; which will replace the existing scale.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-43-1.png" width="70%"></p>
</div>
<div id="automatic-bibliography" class="section level2">
<h2 class="hasAnchor">
<a href="#automatic-bibliography" class="anchor"></a>Automatic bibliography</h2>
<p>Tidybulk provides a handy function called <code>get_bibliography</code> that keeps track of the references for the methods used in your tidybulk workflow. The references are in BibTeX format and can be imported into your reference manager.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="va">counts_de</span><span class="op">)</span>
<span class="co">#&gt;  @Article{tidybulk,</span>
<span class="co">#&gt;   title = {tidybulk: an R tidy framework for modular transcriptomic data analysis},</span>
<span class="co">#&gt;   author = {Stefano Mangiola and Ramyar Molania and Ruining Dong and Maria A. Doyle &amp; Anthony T. Papenfuss},</span>
<span class="co">#&gt;   journal = {Genome Biology},</span>
<span class="co">#&gt;   year = {2021},</span>
<span class="co">#&gt;   volume = {22},</span>
<span class="co">#&gt;   number = {42},</span>
<span class="co">#&gt;   url = {https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7},</span>
<span class="co">#&gt;   }</span>
<span class="co">#&gt; @article{wickham2019welcome,</span>
<span class="co">#&gt;   title={Welcome to the Tidyverse},</span>
<span class="co">#&gt;   author={Wickham, Hadley and Averick, Mara and Bryan, Jennifer and Chang, Winston and McGowan, Lucy D'Agostino and Francois, Romain and Grolemund, Garrett and Hayes, Alex and Henry, Lionel and Hester, Jim and others},</span>
<span class="co">#&gt;   journal={Journal of Open Source Software},</span>
<span class="co">#&gt;   volume={4},</span>
<span class="co">#&gt;   number={43},</span>
<span class="co">#&gt;   pages={1686},</span>
<span class="co">#&gt;   year={2019}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @article{robinson2010edger,</span>
<span class="co">#&gt;   title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},</span>
<span class="co">#&gt;   author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},</span>
<span class="co">#&gt;   journal={Bioinformatics},</span>
<span class="co">#&gt;   volume={26},</span>
<span class="co">#&gt;   number={1},</span>
<span class="co">#&gt;   pages={139--140},</span>
<span class="co">#&gt;   year={2010},</span>
<span class="co">#&gt;   publisher={Oxford University Press}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @article{robinson2010scaling,</span>
<span class="co">#&gt;   title={A scaling normalization method for differential expression analysis of RNA-seq data},</span>
<span class="co">#&gt;   author={Robinson, Mark D and Oshlack, Alicia},</span>
<span class="co">#&gt;   journal={Genome biology},</span>
<span class="co">#&gt;   volume={11},</span>
<span class="co">#&gt;   number={3},</span>
<span class="co">#&gt;   pages={1--9},</span>
<span class="co">#&gt;   year={2010},</span>
<span class="co">#&gt;   publisher={BioMed Central}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @Manual{,</span>
<span class="co">#&gt;     title = {R: A Language and Environment for Statistical Computing},</span>
<span class="co">#&gt;     author = {{R Core Team}},</span>
<span class="co">#&gt;     organization = {R Foundation for Statistical Computing},</span>
<span class="co">#&gt;     address = {Vienna, Austria},</span>
<span class="co">#&gt;     year = {2020},</span>
<span class="co">#&gt;     url = {https://www.R-project.org/},</span>
<span class="co">#&gt;   }</span>
<span class="co">#&gt; @article{lund2012detecting,</span>
<span class="co">#&gt;   title={Detecting differential expression in RNA-sequence data using quasi-likelihood with shrunken dispersion estimates},</span>
<span class="co">#&gt;   author={Lund, Steven P and Nettleton, Dan and McCarthy, Davis J and Smyth, Gordon K},</span>
<span class="co">#&gt;   journal={Statistical applications in genetics and molecular biology},</span>
<span class="co">#&gt;   volume={11},</span>
<span class="co">#&gt;   number={5},</span>
<span class="co">#&gt;   year={2012},</span>
<span class="co">#&gt;   publisher={De Gruyter}</span>
<span class="co">#&gt;     }</span></code></pre></div>
</div>
<div id="cell-type-composition-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-composition-analysis" class="anchor"></a>Cell type composition analysis</h2>
<p>If we are sequencing tissue samples, we may want to know what cell types are present and if there are differences in expression between them. <code>tidybulk</code> has a <code>deconvolve_cellularity</code> function that can help us do this.</p>
<p>For this example, we will use a subset of the breast cancer dataset from <a href="https://www.cancer.gov/tcga">The Cancer Genome Atlas (TCGA)</a>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA</span> <span class="op">&lt;-</span> <span class="fu">rpharma2021tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/BRCA.html">BRCA</a></span> 
<span class="va">BRCA</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 198,374 × 5</span>
<span class="co">#&gt; [90m# Transcripts=9017 | Samples=22 | Assays=count[39m</span>
<span class="co">#&gt;    feature  sample       count  time event_occurred</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt;        &lt;int&gt; &lt;dbl&gt;          &lt;int&gt;</span>
<span class="co">#&gt;  1 A1BG     TCGA-A2-A04P    13   102              1</span>
<span class="co">#&gt;  2 A1BG-AS1 TCGA-A2-A04P    61   102              1</span>
<span class="co">#&gt;  3 A2M      TCGA-A2-A04P 19277   102              1</span>
<span class="co">#&gt;  4 A2MP1    TCGA-A2-A04P     9   102              1</span>
<span class="co">#&gt;  5 AAAS     TCGA-A2-A04P  1917   102              1</span>
<span class="co">#&gt;  6 AACS     TCGA-A2-A04P  1998   102              1</span>
<span class="co">#&gt;  7 AADAT    TCGA-A2-A04P   212   102              1</span>
<span class="co">#&gt;  8 AAGAB    TCGA-A2-A04P  1194   102              1</span>
<span class="co">#&gt;  9 AAK1     TCGA-A2-A04P  1468   102              1</span>
<span class="co">#&gt; 10 AAMP     TCGA-A2-A04P  5798   102              1</span>
<span class="co">#&gt; # … with 40 more rows</span></code></pre></div>
<p>With tidybulk, we can easily infer the proportions of cell types within a tissue using one of several published methods (Cibersort <span class="citation">(Newman et al. 2015)</span>, EPIC <span class="citation">(Racle et al. 2017)</span> and llsr <span class="citation">(Abbas et al. 2009)</span>). Here we will use Cibersort which provides a default signature called LM22 to define the cell types. LM22 contains 547 genes that identify 22 human immune cell types.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type</span> <span class="op">&lt;-</span>
  <span class="va">BRCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>prefix<span class="op">=</span><span class="st">"cibersort: "</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="va">BRCA_cell_type</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 198,374 × 27</span>
<span class="co">#&gt; [90m# Transcripts=9017 | Samples=22 | Assays=count[39m</span>
<span class="co">#&gt;    feature sample count  time event_occurred cibersort..B.ce… cibersort..B.ce…</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt; &lt;dbl&gt;          &lt;int&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt;  1 A1BG    TCGA-…    13   102              1            0.201                0</span>
<span class="co">#&gt;  2 A1BG-A… TCGA-…    61   102              1            0.201                0</span>
<span class="co">#&gt;  3 A2M     TCGA-… 19277   102              1            0.201                0</span>
<span class="co">#&gt;  4 A2MP1   TCGA-…     9   102              1            0.201                0</span>
<span class="co">#&gt;  5 AAAS    TCGA-…  1917   102              1            0.201                0</span>
<span class="co">#&gt;  6 AACS    TCGA-…  1998   102              1            0.201                0</span>
<span class="co">#&gt;  7 AADAT   TCGA-…   212   102              1            0.201                0</span>
<span class="co">#&gt;  8 AAGAB   TCGA-…  1194   102              1            0.201                0</span>
<span class="co">#&gt;  9 AAK1    TCGA-…  1468   102              1            0.201                0</span>
<span class="co">#&gt; 10 AAMP    TCGA-…  5798   102              1            0.201                0</span>
<span class="co">#&gt; # … with 40 more rows, and 20 more variables: cibersort..Plasma.cells &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.CD8 &lt;dbl&gt;, cibersort..T.cells.CD4.naive &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.CD4.memory.resting &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.CD4.memory.activated &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.follicular.helper &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.regulatory..Tregs. &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.gamma.delta &lt;dbl&gt;, cibersort..NK.cells.resting &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..NK.cells.activated &lt;dbl&gt;, cibersort..Monocytes &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Macrophages.M0 &lt;dbl&gt;, cibersort..Macrophages.M1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Macrophages.M2 &lt;dbl&gt;, cibersort..Dendritic.cells.resting &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Dendritic.cells.activated &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Mast.cells.resting &lt;dbl&gt;, cibersort..Mast.cells.activated &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Eosinophils &lt;dbl&gt;, cibersort..Neutrophils &lt;dbl&gt;</span></code></pre></div>
<p>Cell type proportions are added to the tibble as new columns. The prefix makes it easy to reshape the data frame if needed, for visualisation or further analyses.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">&lt;-</span>
  <span class="va">BRCA_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Reshape</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span>
    <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"cibersort"</span><span class="op">)</span>,
    names_prefix <span class="op">=</span> <span class="st">"cibersort: "</span>,
    names_to <span class="op">=</span> <span class="st">"cell_type"</span>,
    values_to <span class="op">=</span> <span class="st">"proportion"</span>
  <span class="op">)</span>

<span class="va">BRCA_cell_type_long</span>
<span class="co">#&gt; # A tibble: 484 × 5</span>
<span class="co">#&gt;    sample        time event_occurred cell_type                        proportion</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;dbl&gt;          &lt;int&gt; &lt;chr&gt;                                 &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-A2-A04P   102              1 cibersort..B.cells.naive             0.201 </span>
<span class="co">#&gt;  2 TCGA-A2-A04P   102              1 cibersort..B.cells.memory            0     </span>
<span class="co">#&gt;  3 TCGA-A2-A04P   102              1 cibersort..Plasma.cells              0     </span>
<span class="co">#&gt;  4 TCGA-A2-A04P   102              1 cibersort..T.cells.CD8               0.0297</span>
<span class="co">#&gt;  5 TCGA-A2-A04P   102              1 cibersort..T.cells.CD4.naive         0     </span>
<span class="co">#&gt;  6 TCGA-A2-A04P   102              1 cibersort..T.cells.CD4.memory.r…     0.0841</span>
<span class="co">#&gt;  7 TCGA-A2-A04P   102              1 cibersort..T.cells.CD4.memory.a…     0     </span>
<span class="co">#&gt;  8 TCGA-A2-A04P   102              1 cibersort..T.cells.follicular.h…     0.0556</span>
<span class="co">#&gt;  9 TCGA-A2-A04P   102              1 cibersort..T.cells.regulatory..…     0.0484</span>
<span class="co">#&gt; 10 TCGA-A2-A04P   102              1 cibersort..T.cells.gamma.delta       0     </span>
<span class="co">#&gt; # … with 474 more rows</span></code></pre></div>
<p>We can plot the proportions of immune cell types for each patient.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot proportions</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
<pre class="challenge_3 challenge"><code>Challenge: What is the most abundant cell type overall in BRCA samples?</code></pre>
</div>
<div id="key-points" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points" class="anchor"></a>Key Points</h2>
<ul>
<li>Bulk RNA sequencing data can be represented and analysed in a ‘tidy’ way using tidySummarizedExperiment, tidybulk and the tidyverse.</li>
<li>tidySummarizedExperiment enables us to visualise and manipulate a Bioconductor SummarizedExperiment object as if it were in tidy data format.</li>
<li>Some of the key steps in an RNA sequencing analysis are filtering lowly abundant transcripts, adjusting for differences in sequencing depth and composition, testing for differential expression, and visualising the data, which can all be performed in a tidy way with tidybulk.</li>
<li>
<code>tidybulk</code> allows streamlined multi-method analyses</li>
<li>
<code>tidybulk</code> allow easy analyses of cell-type composition</li>
</ul>
</div>
</div>
<div id="part-2-single-cell-rna-sequencing-with-tidysinglecellexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#part-2-single-cell-rna-sequencing-with-tidysinglecellexperiment" class="anchor"></a>Part 2 Single-cell RNA sequencing with tidySingleCellExperiment</h1>
<p>A typical single-cell RNA sequencing workflow is shown in the <a href="https://tidytranscriptomics-workshops.github.io/rpharma2021_tidytranscriptomics/articles/tidytranscriptomics.html#differences-between-bulk-and-single-cell-rna-sequencing-1">Workshop Introduction</a> section. We don’t have time in this workshop to go into depth on each step but you can read more about single-cell RNA sequencing workflows in the online book <a href="http://bioconductor.org/books/release/OSCA/index.html">Orchestrating Single-Cell Analysis with Bioconductor</a>.</p>
<p>In Part 1, we showed how we can study the cell-type composition of a biological sample using bulk RNA sequencing. Single-cell sequencing enables a more direct estimation of cell-type composition and gives a greater resolution. For bulk RNA sequencing, we need to infer the cell types using the abundance of transcripts in the whole sample. With single-cell RNA sequencing, we can directly measure the transcripts in each cell and then classify the cells into cell types.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="introduction-to-tidysinglecellexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-tidysinglecellexperiment" class="anchor"></a>Introduction to tidySingleCellExperiment</h2>
<p>The single-cell RNA sequencing data used here is 3000 cells in total, subsetted from 20 samples from 10 peripheral blood mononuclear cell (PBMC) datasets. The datasets are from GSE115189/SRR7244582 <span class="citation">(Freytag et al. 2018)</span>, SRR11038995 [<span class="citation">Cai et al. (2020)</span>, SCP345 (singlecell.broadinstitute.org), SCP424 <span class="citation">(Ding et al. 2020)</span>, SCP591 <span class="citation">(Karagiannis et al. 2020)</span> and 10x-derived 6K and 8K datasets (support.10xgenomics.com/). The data is in <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment</a> format. SingleCellExperiment is a very popular container of single-cell RNA sequencing data.</p>
<p>Similar to what we saw with tidySummarizedExperiment, <code>tidySingleCellExperiment</code> package enables any <em>SingleCellExperiment</em> object to be displayed and manipulated according to tidy data principles without affecting any <em>SingleCellExperiment</em> behaviour.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load pbmc single cell RNA sequencing data</span>
<span class="co"># pbmc &lt;- rpharma2021tidytranscriptomics::pbmc</span>
<span class="co"># </span>
<span class="co"># # take a look</span>
<span class="co"># pbmc</span></code></pre></div>
<p>If we load the <em>tidySingleCellExperiment</em> package and then view the PBMC data, it displays as a tibble.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># library(tidySingleCellExperiment)</span>
<span class="co"># </span>
<span class="co"># pbmc</span></code></pre></div>
<p>It can be interacted with using <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment commands</a> such as <code>assayNames</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># assayNames(pbmc)</span></code></pre></div>
<p>We can also interact with our object as we do with any tidyverse tibble.</p>
<div id="tidyverse-commands-1" class="section level3">
<h3 class="hasAnchor">
<a href="#tidyverse-commands-1" class="anchor"></a>Tidyverse commands</h3>
<p>And now we can also use tidyverse commands, such as <code>filter</code>, <code>select</code> and <code>mutate</code> to explore the tidySingleCellExperiment object. Some examples are shown below and more can be seen at the tidySingleCellExperiment website <a href="https://stemangiola.github.io/tidySingleCellExperiment/articles/introduction.html#tidyverse-commands-1">here</a>. We can also use the tidyverse pipe <code><a href="https://rdrr.io/pkg/tidyseurat/man/pipe.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps. It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<p>We can use <code>filter</code> to choose rows, for example, to see just the rows for the treated samples.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;% filter(ident == "G1")</span></code></pre></div>
<p>We can use <code>select</code> to choose columns, for example, to see the sample, cell, total cellular RNA</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;% select(cell, nCount_RNA , ident)</span></code></pre></div>
<p>We can use <code>mutate</code> to create a column. For example, we could create a new sample_name column that contains shorter sample names. We can remove the SRR1039 prefix that’s present in all of the samples, as shorter names can fit better in some of the plots we will create. We can use <code>mutate</code> together with <code>str_replace</code> to remove the SRR1039 string from the sample column.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;%</span>
<span class="co">#   mutate(ident_l=tolower(ident)) %&gt;%</span>
<span class="co">#   # select columns to view    </span>
<span class="co">#   select(ident, ident_l)</span></code></pre></div>
</div>
<div id="join-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#join-datasets" class="anchor"></a>Join datasets</h3>
<p>We can join datasets as if they were tibbles</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;% bind_rows(pbmc)</span></code></pre></div>
</div>
</div>
<div id="setting-up-the-data-1" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-data-1" class="anchor"></a>Setting up the data</h2>
<p>In this case, we want to polish an annotation column. We will extract the sample, dataset and group information from the file name column into separate columns.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First take a look at the file column</span>
<span class="co"># pbmc %&gt;% select(file)</span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create columns for sample, dataset and groups</span>
<span class="co"># pbmc &lt;-</span>
<span class="co">#   pbmc %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Extract sample and group</span>
<span class="co">#   extract(file, "sample", "../data/([a-zA-Z0-9_]+)/outs.+", remove = FALSE) %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Extract data source</span>
<span class="co">#   extract(file, c("dataset", "groups"), "../data/([a-zA-Z0-9_]+)_([0-9])/outs.+")</span>
<span class="co"># </span>
<span class="co"># # Take a look</span>
<span class="co"># pbmc %&gt;% select(sample, dataset, groups)</span></code></pre></div>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>A key quality control step performed in single-cell analyses is the assessment of the proportion of mitochondrial transcripts. A high mitochondrial count indicates cell death, and it is useful for filtering cells in a dying state.</p>
<p>We’ll first show the mitochondrial analysis for one of the 10 datasets.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># one_dataset &lt;- pbmc %&gt;% filter(dataset =="GSE115189")</span></code></pre></div>
<p>We get the chromosomal location for each gene in the dataset so we can identify the mitochondrial genes. We’ll get a warning that some of the ids don’t find a match, but it should be just a small proportion.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># location &lt;- mapIds(</span>
<span class="co">#   EnsDb.Hsapiens.v86,</span>
<span class="co">#   keys = rownames(one_dataset),</span>
<span class="co">#   column = "SEQNAME",</span>
<span class="co">#   keytype = "SYMBOL"</span>
<span class="co"># )</span></code></pre></div>
<p>Next we calculate the mitchondrial content for each cell in the dataset.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># mito_info_one_dataset &lt;- perCellQCMetrics(one_dataset, subsets = list(Mito = which(location == "MT")))</span>
<span class="co"># </span>
<span class="co"># mito_info_one_dataset</span></code></pre></div>
<p>We then label the cells with high mitochondrial content as outliers.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># mito_info_one_dataset &lt;- mito_info_one_dataset %&gt;%</span>
<span class="co">#       # Converting to tibble</span>
<span class="co">#       as_tibble(rownames = "cell") %&gt;%</span>
<span class="co"># </span>
<span class="co">#       # Label cells with high mitochondrial content</span>
<span class="co">#       mutate(high_mitochondrion = isOutlier(subsets_Mito_percent, type = "higher"))</span>
<span class="co"># </span>
<span class="co"># mito_info_one_dataset</span></code></pre></div>
<p>Finally, we join the mitochondrial information back to the original data so we will be able to filter out the cells with high mitochondrial content.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># mito_info_one_dataset &lt;- one_dataset %&gt;%  left_join(mito_info_one_dataset, by = "cell")</span>
<span class="co"># </span>
<span class="co"># mito_info_one_dataset</span></code></pre></div>
<p>The steps above perform the mitochondrial QC analysis for one dataset. To analyse all 10 datasets, we could repeat the steps for each dataset (a lot of steps!) or create a function and apply it to each dataset. However, a more efficient way is to use nesting. We can nest our data by dataset.</p>
<div id="nesting" class="section level3">
<h3 class="hasAnchor">
<a href="#nesting" class="anchor"></a>Nesting</h3>
<p>A powerful tool tidySingleCellExperiment enables us to use with our single-cell data is tidyverse’s <a href="https://rstudio-education.github.io/tidyverse-cookbook/transform-tables.html#nest-a-data-frame">nest</a>. Nesting allows us to easily perform independent analyses on subsets of the data. For example, our data contains 10 single-cell datasets from different sources, we can nest by dataset and analyse the mitochondrial content for each.</p>
<p>To demonstrate the power of nesting</p>
<p><img src="../inst/vignettes/nesting.png" width="600px"></p>
<div id="examples-of-nesting" class="section level4">
<h4 class="hasAnchor">
<a href="#examples-of-nesting" class="anchor"></a>Examples of nesting</h4>
<p>As a simple example of nesting. Imagine we had two groups, a and b that contained some values.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># my_tibble = </span>
<span class="co">#   tibble(</span>
<span class="co">#   label = c("a","a","a","a","a","b","b","b","b","b" ),</span>
<span class="co">#   value = 1:10</span>
<span class="co"># )</span>
<span class="co"># </span>
<span class="co"># my_tibble</span></code></pre></div>
<p>We could nest this data to make tables containing the values for each of our a and b groups. With <code>nest</code>, <code>data=</code> specifies the columns that will be in the tables and creates a list of these smaller tables.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Nest</span>
<span class="co"># my_tibble_nested = </span>
<span class="co">#   my_tibble %&gt;%</span>
<span class="co">#   nest(data = value)</span>
<span class="co"># </span>
<span class="co"># my_tibble_nested</span></code></pre></div>
<p>We can then iterate over these smaller tables. <code>map</code> is often used in combination with <code>nest</code> to perform commands on nested data. There are a few map commands that you can choose from depending on what you want to return. For example, <code>map_dbl</code> will return a double (floating point number) and we can use it to calculate the average value for each group. You can use <code><a href="https://purrr.tidyverse.org/reference/map.html">?map</a></code> to see the options. The map syntax is <code><a href="https://purrr.tidyverse.org/reference/map.html">map(.x, .f)</a></code> where <code>.x</code> is the table to be iterated over and <code>.f</code> is the function. So in this example, <code>.x</code> is <code>data</code> and <code>.f</code> is <code>~ mean(.x$value)</code>. We get the mean of the value column for each of the nested tables in data.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarise</span>
<span class="co"># my_tibble_nested %&gt;%</span>
<span class="co">#   mutate(average = map_dbl(data, ~ mean(.x$value)))</span></code></pre></div>
<p>We can apply multiple commands to the nested data using the <code><a href="https://rdrr.io/pkg/tidyseurat/man/pipe.html">%&gt;%</a></code>. We can use <code>unnest</code> if we want to get back our large (unnested) table to view the results.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarise + filter</span>
<span class="co"># my_tibble_nested %&gt;%</span>
<span class="co">#   mutate(average = map_dbl(data, ~ mean(.x$value))) %&gt;%</span>
<span class="co">#   filter(average == max(average)) %&gt;%</span>
<span class="co">#   unnest(data)</span></code></pre></div>
<p>We can use <code>map2</code> when we have two arguments that we are iterating over. In the example below we have the nested tables (data) and the average for each (average). The map2 syntax is <code><a href="https://purrr.tidyverse.org/reference/map2.html">map2(.x, .y, .f)</a></code>.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarise + update</span>
<span class="co"># my_tibble_nested %&gt;%</span>
<span class="co">#   mutate(average = map_dbl(data, ~ mean(.x$value))) %&gt;%</span>
<span class="co">#   mutate(data = map2(data, average, ~ .x - .y )) %&gt;%</span>
<span class="co">#   unnest(data)</span></code></pre></div>
<p>We will created nested tables for our 10 datasets.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc_nested &lt;- pbmc %&gt;% </span>
<span class="co">#     nest(data = -sample)</span>
<span class="co"># </span>
<span class="co"># pbmc_nested</span></code></pre></div>
<pre class="challenge_4 challenge"><code>Challenge: Using the nest examples above, can you create a column that includes the mitochondrial information?</code></pre>
<p>We want to obtain this table</p>
<pre><code># A tibble: 10 x 3
   dataset      data             mitochondrion_info
   &lt;chr&gt;        &lt;list&gt;           &lt;list&gt;            
 1 GSE115189    &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 2 10x_6K       &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 3 10x_8K       &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 4 SRR11038995  &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 5 SRR7244582   &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 6 SCP345_580   &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 7 SCP345_860   &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 8 SCP424_pbmc1 &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
 9 SCP424_pbmc2 &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;
10 SCP591       &lt;SnglCllE[,300]&gt; &lt;tibble [300 × 8]&gt;</code></pre>
<p>We will create a table (tibble) with the mitochondrial QC information for each dataset.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># &lt;INSERT THE CODE HERE FROM THE SOLUTIONS&gt;</span>
<span class="co"># mito_info_all_datasets = rpharma2021tidytranscriptomics::mito_info_all_datasets </span></code></pre></div>
<p>As before, we join the mitochondrial information to the original data.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#   # Join the mitochondrial information to the original SingleCellExperiment data</span>
<span class="co">#  mito_info_all_datasets &lt;- mito_info_all_datasets %&gt;%</span>
<span class="co">#     mutate(data = map2(</span>
<span class="co">#     data, mitochondrion_info,</span>
<span class="co">#     ~ left_join(.x, .y, by = "cell")</span>
<span class="co">#   )) %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Remove the separate mitochondrial information table</span>
<span class="co">#   select(-mitochondrion_info)</span>
<span class="co"># </span>
<span class="co"># mito_info_all_datasets</span></code></pre></div>
<p>Remove the nesting to get back one table containing all datasets.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc &lt;- mito_info_all_datasets %&gt;%</span>
<span class="co">#   unnest(data)</span>
<span class="co"># </span>
<span class="co"># pbmc</span></code></pre></div>
<p>Nesting has allowed us to perform the QC on each dataset in an efficient way.</p>
<p>We can use tidyverse to reshape the data and create <a href="http://www.cbs.dtu.dk/~eklund/beeswarm/">beeswarm plots</a> to visualise the mitochondrial content.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Reshaping</span>
<span class="co">#   pivot_longer(c(detected, sum, subsets_Mito_percent)) %&gt;%</span>
<span class="co">#   ggplot(aes(</span>
<span class="co">#     x = sample, y = value,</span>
<span class="co">#     color = high_mitochondrion,</span>
<span class="co">#     alpha = high_mitochondrion,</span>
<span class="co">#     size = high_mitochondrion</span>
<span class="co">#   )) +</span>
<span class="co"># </span>
<span class="co">#   # Plotting</span>
<span class="co">#   geom_quasirandom() +</span>
<span class="co">#   facet_wrap(~name, scale = "free_y") +</span>
<span class="co"># </span>
<span class="co">#   # Customisation</span>
<span class="co">#   scale_color_manual(values = c("black", "#e11f28")) +</span>
<span class="co">#   scale_size_discrete(range = c(0, 2)) +</span>
<span class="co">#   theme_bw() +</span>
<span class="co">#   theme(axis.text.x = element_text(angle = 50, hjust = 1, vjust = 1))</span></code></pre></div>
<p>In the faceted plot, “detected” is the number of genes in each of the 10 datasets, “sum” is the total counts.</p>
<p>We then proceed to filter out cells with high mitochondrial content.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc &lt;- pbmc %&gt;% filter(!high_mitochondrion)</span></code></pre></div>
</div>
</div>
</div>
<div id="scaling-and-integrating" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-and-integrating" class="anchor"></a>Scaling and Integrating</h2>
<p>As we are working with multiple datasets, we need to integrate them and adjust for technical variability between them. Here we’ll nest by dataset (batch), normalise within each batch with <code>multiBatchNorm</code> and correct for batch effects with <code>fastMNN</code>.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scaling within each dataset</span>
<span class="co"># pbmc &lt;-</span>
<span class="co">#   pbmc %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Define batch</span>
<span class="co">#   nest(data = -dataset) %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Normalisation   </span>
<span class="co">#   mutate(data = multiBatchNorm(data)) %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Integration</span>
<span class="co">#   pull(data) %&gt;%</span>
<span class="co">#   fastMNN() %&gt;%</span>
<span class="co">#   </span>
<span class="co">#   # Join old information</span>
<span class="co">#   left_join(pbmc %&gt;% as_tibble())</span></code></pre></div>
</div>
<div id="identify-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h2>
<p>We proceed with identifying cell clusters.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Assign clusters to the 'colLabels'</span>
<span class="co"># of the SingleCellExperiment object</span>
<span class="co"># colLabels(pbmc) &lt;- # from SingleCellExperiment</span>
<span class="co">#   pbmc %&gt;%</span>
<span class="co">#   buildSNNGraph(use.dimred="corrected") %&gt;% # from scran - shared nearest neighbour</span>
<span class="co">#   cluster_walktrap() %$% # from igraph</span>
<span class="co">#   membership %&gt;%</span>
<span class="co">#   as.factor()</span>
<span class="co"># </span>
<span class="co"># # Reorder columns</span>
<span class="co"># pbmc %&gt;% select(label, everything())</span></code></pre></div>
<p>Thanks to tidySingleCellExperiment we can interrogate the object with tidyverse commands and use <code>count</code> to count the number of cells in each cluster.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;% count(label)</span></code></pre></div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h2>
<p>Besides PCA which is a linear dimensionality reduction, we can apply neighbour aware methods such as UMAP, to better define locally similar cells.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate UMAP with scater</span>
<span class="co"># pbmc %&gt;%</span>
<span class="co"># runUMAP(ncomponents = 2, dimred="corrected") # from scater</span></code></pre></div>
<pre class="challenge_5 challenge"><code>Without creating any variable, check if the variability explained by the first UMAP dimension when calculating 2 components (ncomponents = 2) is equal/more/less than when calculating 3 components.</code></pre>
<p><strong>Tip: your can use as_tibble() to convert the tibble abstraction to a simple (and light) cell-wise tibble</strong></p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># &lt;INSERT THE CODE HERE FROM THE SOLUTIONS&gt;</span></code></pre></div>
<p>We can calculate the first 3 UMAP dimensions using the scater framework.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate UMAP with scater</span>
<span class="co"># pbmc &lt;-</span>
<span class="co">#   pbmc %&gt;%</span>
<span class="co">#   runUMAP(ncomponents = 3, dimred="corrected") # from scater</span></code></pre></div>
<p>And we can plot the clusters as a 3D plot using plotly. This time we are colouring by estimated cluster labels to visually check the cluster labels.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc %&gt;%</span>
<span class="co">#   plot_ly(</span>
<span class="co">#     x = ~`UMAP1`,</span>
<span class="co">#     y = ~`UMAP2`,</span>
<span class="co">#     z = ~`UMAP3`,</span>
<span class="co">#     colors = friendly_cols[1:10],</span>
<span class="co">#     color = ~label,</span>
<span class="co">#     size=0.5 </span>
<span class="co">#   )</span></code></pre></div>
<p><img src="../inst/vignettes/plotly_2.png" width="261"></p>
</div>
<div id="cell-type-classification" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-classification" class="anchor"></a>Cell type classification</h2>
<div id="manual-cell-type-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#manual-cell-type-classification" class="anchor"></a>Manual cell type classification</h3>
<p>We can identify cluster markers (genes). As example, we are selecting the top 10 for each cluster. We can then plot a heatmap of those gene markers across cells.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Identify top 10 markers per cluster</span>
<span class="co"># marker_genes &lt;-</span>
<span class="co">#     pbmc %&gt;%</span>
<span class="co">#     findMarkers(groups=pbmc$label, assay.type = "reconstructed") %&gt;%  # scran</span>
<span class="co">#     as.list() %&gt;%</span>
<span class="co">#     map(~ head(.x, 10) %&gt;%  rownames()) %&gt;%</span>
<span class="co">#     unlist()</span>
<span class="co"># </span>
<span class="co"># # Plot heatmap</span>
<span class="co"># pbmc %&gt;%</span>
<span class="co">#   plotHeatmap(                                  # from scater</span>
<span class="co">#     features=marker_genes, </span>
<span class="co">#     columns=order(pbmc$label), </span>
<span class="co">#     colour_columns_by=c("label"), </span>
<span class="co">#     exprs_values  = "reconstructed"</span>
<span class="co">#   ) </span></code></pre></div>
</div>
<div id="automatic-cell-type-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#automatic-cell-type-classification" class="anchor"></a>Automatic cell type classification</h3>
<p>We can infer cell type identities (e.g. T cell) using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output using tidyverse. SingleR <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html">accepts any log-normalised transcript abundance matrix</a></p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reference cell types</span>
<span class="co"># blueprint &lt;- BlueprintEncodeData()</span>
<span class="co"># </span>
<span class="co"># cell_type &lt;-</span>
<span class="co"># </span>
<span class="co">#   # extracting counts from SingleCellExperiment object</span>
<span class="co">#   assays(pbmc)$reconstructed %&gt;%</span>
<span class="co"># </span>
<span class="co">#     # SingleR</span>
<span class="co">#   SingleR(</span>
<span class="co">#     ref = blueprint,</span>
<span class="co">#     labels = blueprint$label.main,</span>
<span class="co">#     clusters = pbmc %&gt;% pull(label)</span>
<span class="co">#   ) %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Formatting results</span>
<span class="co">#   as.data.frame() %&gt;%</span>
<span class="co">#   as_tibble(rownames = "label") %&gt;%</span>
<span class="co">#   select(label, first.labels)</span>
<span class="co"># </span>
<span class="co"># cell_type</span></code></pre></div>
<p>We join the cell type information to our pbmc data.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Join cell type info</span>
<span class="co"># pbmc &lt;-</span>
<span class="co">#   pbmc %&gt;%</span>
<span class="co">#   left_join(cell_type, by = "label")</span>
<span class="co"># </span>
<span class="co"># # Reorder columns</span>
<span class="co"># pbmc %&gt;% select(cell, first.labels, everything())</span></code></pre></div>
<div id="challenge-use-tidyverse-grammar-to-identify-the-most-abundant-cell-type" class="section level4">
<h4 class="hasAnchor">
<a href="#challenge-use-tidyverse-grammar-to-identify-the-most-abundant-cell-type" class="anchor"></a>Challenge: use tidyverse grammar to identify the most abundant cell type</h4>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># &lt;INSERT THE CODE HERE FROM THE SOLUTIONS&gt;</span></code></pre></div>
</div>
</div>
</div>
<div id="pseudobulk-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#pseudobulk-analyses" class="anchor"></a>Pseudobulk analyses</h2>
<p>It is sometime useful to aggregate cell-wise transcript abundance into pseudobulk samples. It is possible to explore data and perform hypothesis testing with tools and data-source that we are more familiar with. For example, we can use edgeR in tidybulk to perform differential expression testing. For more details on pseudobulk analysis see <a href="https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html">here</a>.</p>
<div id="data-exploration-using-pseudobulk-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#data-exploration-using-pseudobulk-samples" class="anchor"></a>Data exploration using pseudobulk samples</h3>
<p>To do this, we will load a helper function called <code>aggregate_cells</code> from <a href="https://gist.github.com/stemangiola/ebce5a3f931b298611b56680d39c073c">GitHub Gist</a> and use it to create a group for each sample.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load aggregate function</span>
<span class="co"># devtools::source_gist("889e4089a0e5ca2649d3c9164e3b542e")</span>
<span class="co"># </span>
<span class="co"># # Aggregate</span>
<span class="co"># pbmc_bulk =</span>
<span class="co">#   rpharma2021tidytranscriptomics::pbmc %&gt;%</span>
<span class="co">#   aggregate_cells(file)</span>
<span class="co"># pbmc_bulk</span></code></pre></div>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc_bulk %&gt;%</span>
<span class="co"># </span>
<span class="co">#   # Tidybulk operations</span>
<span class="co">#   tidybulk::identify_abundant() %&gt;%</span>
<span class="co">#   tidybulk::scale_abundance()</span></code></pre></div>
</div>
</div>
<div id="introduction-to-tidyseurat" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-tidyseurat" class="anchor"></a>Introduction to tidyseurat</h2>
<p>In Part 2 we showed how we can study the cell-type composition of a biological sample using bulk RNA sequencing. Single cell sequencing enables a more direct estimation of cell-type composition and gives greater resolution. For bulk RNA sequencing we need to infer the cell types using the abundance of transcripts in the whole sample, with single-cell RNA sequencing we can directly measure the transcripts in each cell and then classify the cells into cell types.</p>
<p>Seurat is a very popular analysis toolkit for single cell RNA sequencing data <span class="citation">(Butler et al. 2018; Stuart et al. 2019)</span> .</p>
<p>tidyseurat provides a bridge between the Seurat single-cell package <span class="citation">(Butler et al. 2018; Stuart et al. 2019)</span> and the tidyverse <span class="citation">(Wickham et al. 2019)</span>. It creates an invisible layer that enables viewing the Seurat object as a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_seurat</span> <span class="op">&lt;-</span> <span class="fu">rpharma2021tidytranscriptomics</span><span class="fu">::</span><span class="va">pbmc_seurat</span>
<span class="co">#&gt; Registered S3 method overwritten by 'spatstat.geom':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   print.boxx cli</span></code></pre></div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidyseurat">tidyseurat</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_seurat</span>
<span class="co">#&gt; # A Seurat-tibble abstraction: 80 × 9</span>
<span class="co">#&gt; [90m# Features=230 | Active assay=RNA | Assays=RNA[39m</span>
<span class="co">#&gt;    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span>
<span class="co">#&gt;  1 ATGC… SeuratPro…         70           47 0               A             g2    </span>
<span class="co">#&gt;  2 CATG… SeuratPro…         85           52 0               A             g1    </span>
<span class="co">#&gt;  3 GAAC… SeuratPro…         87           50 1               B             g2    </span>
<span class="co">#&gt;  4 TGAC… SeuratPro…        127           56 0               A             g2    </span>
<span class="co">#&gt;  5 AGTC… SeuratPro…        173           53 0               A             g2    </span>
<span class="co">#&gt;  6 TCTG… SeuratPro…         70           48 0               A             g1    </span>
<span class="co">#&gt;  7 TGGT… SeuratPro…         64           36 0               A             g1    </span>
<span class="co">#&gt;  8 GCAG… SeuratPro…         72           45 0               A             g1    </span>
<span class="co">#&gt;  9 GATA… SeuratPro…         52           36 0               A             g1    </span>
<span class="co">#&gt; 10 AATG… SeuratPro…        100           41 0               A             g1    </span>
<span class="co">#&gt; # … with 70 more rows, and 2 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;</span></code></pre></div>
<p>It can be interacted with using <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment commands</a> such as <code>assayNames</code>.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Assays</span><span class="op">(</span><span class="va">pbmc_seurat</span><span class="op">)</span>
<span class="co">#&gt; [1] "RNA"</span></code></pre></div>
<p>We can also interact with our object as we do with any tidyverse tibble.</p>
<div id="tidyverse-commands-2" class="section level3">
<h3 class="hasAnchor">
<a href="#tidyverse-commands-2" class="anchor"></a>Tidyverse commands</h3>
<p>We can interact to this object in the same way we interact with <code>SingleCellExperiment</code>, through the tidy paradigm</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filter</span>
<span class="va">pbmc_seurat</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">groups</span> <span class="op">==</span> <span class="st">"g1"</span><span class="op">)</span>
<span class="co">#&gt; # A Seurat-tibble abstraction: 44 × 9</span>
<span class="co">#&gt; [90m# Features=230 | Active assay=RNA | Assays=RNA[39m</span>
<span class="co">#&gt;    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; </span>
<span class="co">#&gt;  1 CATG… SeuratPro…         85           52 0               A             g1    </span>
<span class="co">#&gt;  2 TCTG… SeuratPro…         70           48 0               A             g1    </span>
<span class="co">#&gt;  3 TGGT… SeuratPro…         64           36 0               A             g1    </span>
<span class="co">#&gt;  4 GCAG… SeuratPro…         72           45 0               A             g1    </span>
<span class="co">#&gt;  5 GATA… SeuratPro…         52           36 0               A             g1    </span>
<span class="co">#&gt;  6 AATG… SeuratPro…        100           41 0               A             g1    </span>
<span class="co">#&gt;  7 AGAG… SeuratPro…        191           61 0               A             g1    </span>
<span class="co">#&gt;  8 CTAA… SeuratPro…        168           44 0               A             g1    </span>
<span class="co">#&gt;  9 TTGG… SeuratPro…        135           45 0               A             g1    </span>
<span class="co">#&gt; 10 CATC… SeuratPro…         79           43 0               A             g1    </span>
<span class="co">#&gt; # … with 34 more rows, and 2 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;</span>

<span class="co"># Select</span>
<span class="va">pbmc_seurat</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">nCount_RNA</span> , <span class="va">groups</span><span class="op">)</span>
<span class="co">#&gt; # A Seurat-tibble abstraction: 80 × 3</span>
<span class="co">#&gt; [90m# Features=230 | Active assay=RNA | Assays=RNA[39m</span>
<span class="co">#&gt;    cell           nCount_RNA groups</span>
<span class="co">#&gt;    &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt; </span>
<span class="co">#&gt;  1 ATGCCAGAACGACT         70 g2    </span>
<span class="co">#&gt;  2 CATGGCCTGTGCAT         85 g1    </span>
<span class="co">#&gt;  3 GAACCTGATGAACC         87 g2    </span>
<span class="co">#&gt;  4 TGACTGGATTCTCA        127 g2    </span>
<span class="co">#&gt;  5 AGTCAGACTGCACA        173 g2    </span>
<span class="co">#&gt;  6 TCTGATACACGTGT         70 g1    </span>
<span class="co">#&gt;  7 TGGTATCTAAACAG         64 g1    </span>
<span class="co">#&gt;  8 GCAGCTCTGTTTCT         72 g1    </span>
<span class="co">#&gt;  9 GATATAACACGCAT         52 g1    </span>
<span class="co">#&gt; 10 AATGTTGACAGTCA        100 g1    </span>
<span class="co">#&gt; # … with 70 more rows</span>

<span class="co"># Mutate</span>
<span class="va">pbmc_seurat</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>groups_l<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html">tolower</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># select columns to view    </span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">groups_l</span><span class="op">)</span>
<span class="co">#&gt; tidyseurat says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 80 × 2</span>
<span class="co">#&gt;    groups groups_l</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;   </span>
<span class="co">#&gt;  1 g2     g2      </span>
<span class="co">#&gt;  2 g1     g1      </span>
<span class="co">#&gt;  3 g2     g2      </span>
<span class="co">#&gt;  4 g2     g2      </span>
<span class="co">#&gt;  5 g2     g2      </span>
<span class="co">#&gt;  6 g1     g1      </span>
<span class="co">#&gt;  7 g1     g1      </span>
<span class="co">#&gt;  8 g1     g1      </span>
<span class="co">#&gt;  9 g1     g1      </span>
<span class="co">#&gt; 10 g1     g1      </span>
<span class="co">#&gt; # … with 70 more rows</span>

<span class="co"># Bind datasets</span>
<span class="va">pbmc_seurat</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidyseurat/man/dplyr-methods.html">bind_rows</a></span><span class="op">(</span><span class="va">pbmc_seurat</span><span class="op">)</span>
<span class="co">#&gt; Warning in CheckDuplicateCellNames(object.list = objects): Some cell names are</span>
<span class="co">#&gt; duplicated across objects provided. Renaming to enforce unique cell names.</span>
<span class="co">#&gt; # A Seurat-tibble abstraction: 160 × 9</span>
<span class="co">#&gt; [90m# Features=230 | Active assay=RNA | Assays=RNA[39m</span>
<span class="co">#&gt;    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;chr&gt;           &lt;chr&gt;         &lt;chr&gt; </span>
<span class="co">#&gt;  1 ATGC… SeuratPro…         70           47 0               A             g2    </span>
<span class="co">#&gt;  2 CATG… SeuratPro…         85           52 0               A             g1    </span>
<span class="co">#&gt;  3 GAAC… SeuratPro…         87           50 1               B             g2    </span>
<span class="co">#&gt;  4 TGAC… SeuratPro…        127           56 0               A             g2    </span>
<span class="co">#&gt;  5 AGTC… SeuratPro…        173           53 0               A             g2    </span>
<span class="co">#&gt;  6 TCTG… SeuratPro…         70           48 0               A             g1    </span>
<span class="co">#&gt;  7 TGGT… SeuratPro…         64           36 0               A             g1    </span>
<span class="co">#&gt;  8 GCAG… SeuratPro…         72           45 0               A             g1    </span>
<span class="co">#&gt;  9 GATA… SeuratPro…         52           36 0               A             g1    </span>
<span class="co">#&gt; 10 AATG… SeuratPro…        100           41 0               A             g1    </span>
<span class="co">#&gt; # … with 150 more rows, and 2 more variables: RNA_snn_res.1 &lt;chr&gt;, file &lt;chr&gt;</span></code></pre></div>
</div>
</div>
<div id="key-points-1" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points-1" class="anchor"></a>Key Points</h2>
<ul>
<li>Some basic steps of a single-cell RNA sequencing analysis are dimensionality reduction, cluster identification and cell type classification</li>
<li>
<code>tidySingleCellExperiment</code> is an invisible layer that operates on a <code>SingleCellExperiment</code> object and enables us to visualise and manipulate data as if it were a tidy data frame.</li>
<li>
<code>tidySingleCellExperiment</code> object is a <code>SingleCellExperiment object</code> so it can be used with any <code>SingleCellExperiment</code> compatible method</li>
</ul>
</div>
</div>
<div id="feedback" class="section level1">
<h1 class="hasAnchor">
<a href="#feedback" class="anchor"></a>Feedback</h1>
<p>We would be very grateful if you could please complete the feedback survey so we can gather feedback on the effectiveness and usefulness of this tutorial. The link to the survey is here:</p>
</div>
<div id="contributing" class="section level1">
<h1 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h1>
<p>If you want to suggest improvements for this workshop or ask questions, you can do so as described <a href="https://github.com/tidytranscriptomics-workshops/rpharma2021_tidytranscriptomics/blob/master/CONTRIBUTING.md">here</a>.</p>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>Record package and version information with <code>sessionInfo</code></p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.1 (2021-08-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] tidyseurat_0.3.0               SeuratObject_4.0.2            </span>
<span class="co">#&gt;  [3] ggbeeswarm_0.6.0               celldex_1.3.0                 </span>
<span class="co">#&gt;  [5] EnsDb.Hsapiens.v86_2.99.0      ensembldb_2.17.4              </span>
<span class="co">#&gt;  [7] AnnotationFilter_1.17.1        GenomicFeatures_1.45.2        </span>
<span class="co">#&gt;  [9] AnnotationDbi_1.55.2           SingleR_1.7.1                 </span>
<span class="co">#&gt; [11] batchelor_1.9.1                igraph_1.2.7                  </span>
<span class="co">#&gt; [13] scran_1.21.3                   scater_1.21.8                 </span>
<span class="co">#&gt; [15] scuttle_1.3.1                  SingleCellExperiment_1.15.2   </span>
<span class="co">#&gt; [17] tidySummarizedExperiment_1.3.0 tidybulk_1.5.5                </span>
<span class="co">#&gt; [19] tidyHeatmap_1.3.1              GGally_2.1.2                  </span>
<span class="co">#&gt; [21] ggrepel_0.9.1                  plotly_4.10.0                 </span>
<span class="co">#&gt; [23] purrr_0.3.4                    ggplot2_3.3.5                 </span>
<span class="co">#&gt; [25] stringr_1.4.0                  readr_2.0.2                   </span>
<span class="co">#&gt; [27] tidyr_1.1.4                    dplyr_1.0.7                   </span>
<span class="co">#&gt; [29] tibble_3.1.5                   airway_1.13.0                 </span>
<span class="co">#&gt; [31] SummarizedExperiment_1.23.5    Biobase_2.53.0                </span>
<span class="co">#&gt; [33] GenomicRanges_1.45.0           GenomeInfoDb_1.29.10          </span>
<span class="co">#&gt; [35] IRanges_2.27.2                 S4Vectors_0.31.5              </span>
<span class="co">#&gt; [37] BiocGenerics_0.39.2            MatrixGenerics_1.5.4          </span>
<span class="co">#&gt; [39] matrixStats_0.61.0            </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] rappdirs_0.3.3                       </span>
<span class="co">#&gt;   [2] rtracklayer_1.53.1                   </span>
<span class="co">#&gt;   [3] scattermore_0.7                      </span>
<span class="co">#&gt;   [4] ragg_1.1.3                           </span>
<span class="co">#&gt;   [5] bit64_4.0.5                          </span>
<span class="co">#&gt;   [6] knitr_1.36                           </span>
<span class="co">#&gt;   [7] irlba_2.3.3                          </span>
<span class="co">#&gt;   [8] DelayedArray_0.19.4                  </span>
<span class="co">#&gt;   [9] rpart_4.1-15                         </span>
<span class="co">#&gt;  [10] data.table_1.14.2                    </span>
<span class="co">#&gt;  [11] KEGGREST_1.33.0                      </span>
<span class="co">#&gt;  [12] RCurl_1.98-1.5                       </span>
<span class="co">#&gt;  [13] doParallel_1.0.16                    </span>
<span class="co">#&gt;  [14] generics_0.1.1                       </span>
<span class="co">#&gt;  [15] preprocessCore_1.55.2                </span>
<span class="co">#&gt;  [16] ScaledMatrix_1.1.0                   </span>
<span class="co">#&gt;  [17] cowplot_1.1.1                        </span>
<span class="co">#&gt;  [18] RSQLite_2.2.8                        </span>
<span class="co">#&gt;  [19] RANN_2.6.1                           </span>
<span class="co">#&gt;  [20] proxy_0.4-26                         </span>
<span class="co">#&gt;  [21] future_1.22.1                        </span>
<span class="co">#&gt;  [22] bit_4.0.4                            </span>
<span class="co">#&gt;  [23] tzdb_0.1.2                           </span>
<span class="co">#&gt;  [24] spatstat.data_2.1-0                  </span>
<span class="co">#&gt;  [25] xml2_1.3.2                           </span>
<span class="co">#&gt;  [26] httpuv_1.6.3                         </span>
<span class="co">#&gt;  [27] assertthat_0.2.1                     </span>
<span class="co">#&gt;  [28] viridis_0.6.2                        </span>
<span class="co">#&gt;  [29] xfun_0.27                            </span>
<span class="co">#&gt;  [30] hms_1.1.1                            </span>
<span class="co">#&gt;  [31] jquerylib_0.1.4                      </span>
<span class="co">#&gt;  [32] evaluate_0.14                        </span>
<span class="co">#&gt;  [33] promises_1.2.0.1                     </span>
<span class="co">#&gt;  [34] fansi_0.5.0                          </span>
<span class="co">#&gt;  [35] restfulr_0.0.13                      </span>
<span class="co">#&gt;  [36] progress_1.2.2                       </span>
<span class="co">#&gt;  [37] dendextend_1.15.1                    </span>
<span class="co">#&gt;  [38] dbplyr_2.1.1                         </span>
<span class="co">#&gt;  [39] DBI_1.1.1                            </span>
<span class="co">#&gt;  [40] geneplotter_1.71.0                   </span>
<span class="co">#&gt;  [41] htmlwidgets_1.5.4                    </span>
<span class="co">#&gt;  [42] reshape_0.8.8                        </span>
<span class="co">#&gt;  [43] spatstat.geom_2.3-0                  </span>
<span class="co">#&gt;  [44] ellipsis_0.3.2                       </span>
<span class="co">#&gt;  [45] annotate_1.71.0                      </span>
<span class="co">#&gt;  [46] deldir_1.0-6                         </span>
<span class="co">#&gt;  [47] biomaRt_2.49.7                       </span>
<span class="co">#&gt;  [48] sparseMatrixStats_1.5.3              </span>
<span class="co">#&gt;  [49] vctrs_0.3.8                          </span>
<span class="co">#&gt;  [50] ROCR_1.0-11                          </span>
<span class="co">#&gt;  [51] abind_1.4-5                          </span>
<span class="co">#&gt;  [52] cachem_1.0.6                         </span>
<span class="co">#&gt;  [53] withr_2.4.2                          </span>
<span class="co">#&gt;  [54] sctransform_0.3.2                    </span>
<span class="co">#&gt;  [55] GenomicAlignments_1.29.0             </span>
<span class="co">#&gt;  [56] prettyunits_1.1.1                    </span>
<span class="co">#&gt;  [57] dittoSeq_1.5.4                       </span>
<span class="co">#&gt;  [58] goftest_1.2-3                        </span>
<span class="co">#&gt;  [59] cluster_2.1.2                        </span>
<span class="co">#&gt;  [60] ExperimentHub_2.1.4                  </span>
<span class="co">#&gt;  [61] lazyeval_0.2.2                       </span>
<span class="co">#&gt;  [62] crayon_1.4.1                         </span>
<span class="co">#&gt;  [63] genefilter_1.75.1                    </span>
<span class="co">#&gt;  [64] edgeR_3.35.3                         </span>
<span class="co">#&gt;  [65] pkgconfig_2.0.3                      </span>
<span class="co">#&gt;  [66] labeling_0.4.2                       </span>
<span class="co">#&gt;  [67] nlme_3.1-153                         </span>
<span class="co">#&gt;  [68] vipor_0.4.5                          </span>
<span class="co">#&gt;  [69] ProtGenerics_1.25.1                  </span>
<span class="co">#&gt;  [70] rlang_0.4.12                         </span>
<span class="co">#&gt;  [71] globals_0.14.0                       </span>
<span class="co">#&gt;  [72] miniUI_0.1.1.1                       </span>
<span class="co">#&gt;  [73] lifecycle_1.0.1                      </span>
<span class="co">#&gt;  [74] filelock_1.0.2                       </span>
<span class="co">#&gt;  [75] BiocFileCache_2.1.1                  </span>
<span class="co">#&gt;  [76] rsvd_1.0.5                           </span>
<span class="co">#&gt;  [77] AnnotationHub_3.1.7                  </span>
<span class="co">#&gt;  [78] polyclip_1.10-0                      </span>
<span class="co">#&gt;  [79] rprojroot_2.0.2                      </span>
<span class="co">#&gt;  [80] lmtest_0.9-38                        </span>
<span class="co">#&gt;  [81] Matrix_1.3-4                         </span>
<span class="co">#&gt;  [82] zoo_1.8-9                            </span>
<span class="co">#&gt;  [83] beeswarm_0.4.0                       </span>
<span class="co">#&gt;  [84] ggridges_0.5.3                       </span>
<span class="co">#&gt;  [85] GlobalOptions_0.1.2                  </span>
<span class="co">#&gt;  [86] pheatmap_1.0.12                      </span>
<span class="co">#&gt;  [87] png_0.1-7                            </span>
<span class="co">#&gt;  [88] viridisLite_0.4.0                    </span>
<span class="co">#&gt;  [89] rjson_0.2.20                         </span>
<span class="co">#&gt;  [90] bitops_1.0-7                         </span>
<span class="co">#&gt;  [91] KernSmooth_2.23-20                   </span>
<span class="co">#&gt;  [92] Biostrings_2.61.2                    </span>
<span class="co">#&gt;  [93] blob_1.2.2                           </span>
<span class="co">#&gt;  [94] DelayedMatrixStats_1.15.4            </span>
<span class="co">#&gt;  [95] shape_1.4.6                          </span>
<span class="co">#&gt;  [96] parallelly_1.28.1                    </span>
<span class="co">#&gt;  [97] beachmat_2.9.1                       </span>
<span class="co">#&gt;  [98] scales_1.1.1                         </span>
<span class="co">#&gt;  [99] memoise_2.0.0                        </span>
<span class="co">#&gt; [100] magrittr_2.0.1                       </span>
<span class="co">#&gt; [101] plyr_1.8.6                           </span>
<span class="co">#&gt; [102] ica_1.0-2                            </span>
<span class="co">#&gt; [103] zlibbioc_1.39.0                      </span>
<span class="co">#&gt; [104] compiler_4.1.1                       </span>
<span class="co">#&gt; [105] dqrng_0.3.0                          </span>
<span class="co">#&gt; [106] BiocIO_1.3.0                         </span>
<span class="co">#&gt; [107] RColorBrewer_1.1-2                   </span>
<span class="co">#&gt; [108] clue_0.3-60                          </span>
<span class="co">#&gt; [109] DESeq2_1.33.5                        </span>
<span class="co">#&gt; [110] fitdistrplus_1.1-6                   </span>
<span class="co">#&gt; [111] Rsamtools_2.9.1                      </span>
<span class="co">#&gt; [112] cli_3.0.1                            </span>
<span class="co">#&gt; [113] XVector_0.33.0                       </span>
<span class="co">#&gt; [114] listenv_0.8.0                        </span>
<span class="co">#&gt; [115] pbapply_1.5-0                        </span>
<span class="co">#&gt; [116] patchwork_1.1.1                      </span>
<span class="co">#&gt; [117] mgcv_1.8-38                          </span>
<span class="co">#&gt; [118] MASS_7.3-54                          </span>
<span class="co">#&gt; [119] tidyselect_1.1.1                     </span>
<span class="co">#&gt; [120] stringi_1.7.5                        </span>
<span class="co">#&gt; [121] textshaping_0.3.6                    </span>
<span class="co">#&gt; [122] highr_0.9                            </span>
<span class="co">#&gt; [123] yaml_2.2.1                           </span>
<span class="co">#&gt; [124] BiocSingular_1.9.1                   </span>
<span class="co">#&gt; [125] locfit_1.5-9.4                       </span>
<span class="co">#&gt; [126] grid_4.1.1                           </span>
<span class="co">#&gt; [127] sass_0.4.0                           </span>
<span class="co">#&gt; [128] tools_4.1.1                          </span>
<span class="co">#&gt; [129] future.apply_1.8.1                   </span>
<span class="co">#&gt; [130] parallel_4.1.1                       </span>
<span class="co">#&gt; [131] circlize_0.4.13                      </span>
<span class="co">#&gt; [132] bluster_1.3.2                        </span>
<span class="co">#&gt; [133] foreach_1.5.1                        </span>
<span class="co">#&gt; [134] metapod_1.1.0                        </span>
<span class="co">#&gt; [135] gridExtra_2.3                        </span>
<span class="co">#&gt; [136] farver_2.1.0                         </span>
<span class="co">#&gt; [137] Rtsne_0.15                           </span>
<span class="co">#&gt; [138] digest_0.6.28                        </span>
<span class="co">#&gt; [139] BiocManager_1.30.16                  </span>
<span class="co">#&gt; [140] shiny_1.7.1                          </span>
<span class="co">#&gt; [141] Rcpp_1.0.7                           </span>
<span class="co">#&gt; [142] BiocVersion_3.14.0                   </span>
<span class="co">#&gt; [143] later_1.3.0                          </span>
<span class="co">#&gt; [144] RcppAnnoy_0.0.19                     </span>
<span class="co">#&gt; [145] org.Hs.eg.db_3.14.0                  </span>
<span class="co">#&gt; [146] httr_1.4.2                           </span>
<span class="co">#&gt; [147] rpharma2021tidytranscriptomics_0.11.0</span>
<span class="co">#&gt; [148] ComplexHeatmap_2.9.4                 </span>
<span class="co">#&gt; [149] colorspace_2.0-2                     </span>
<span class="co">#&gt; [150] tensor_1.5                           </span>
<span class="co">#&gt; [151] reticulate_1.22                      </span>
<span class="co">#&gt; [152] XML_3.99-0.8                         </span>
<span class="co">#&gt; [153] fs_1.5.0                             </span>
<span class="co">#&gt; [154] splines_4.1.1                        </span>
<span class="co">#&gt; [155] uwot_0.1.10                          </span>
<span class="co">#&gt; [156] statmod_1.4.36                       </span>
<span class="co">#&gt; [157] spatstat.utils_2.2-0                 </span>
<span class="co">#&gt; [158] pkgdown_1.6.1                        </span>
<span class="co">#&gt; [159] systemfonts_1.0.3                    </span>
<span class="co">#&gt; [160] xtable_1.8-4                         </span>
<span class="co">#&gt; [161] jsonlite_1.7.2                       </span>
<span class="co">#&gt; [162] R6_2.5.1                             </span>
<span class="co">#&gt; [163] pillar_1.6.4                         </span>
<span class="co">#&gt; [164] htmltools_0.5.2                      </span>
<span class="co">#&gt; [165] mime_0.12                            </span>
<span class="co">#&gt; [166] glue_1.4.2                           </span>
<span class="co">#&gt; [167] fastmap_1.1.0                        </span>
<span class="co">#&gt; [168] BiocParallel_1.27.17                 </span>
<span class="co">#&gt; [169] BiocNeighbors_1.11.0                 </span>
<span class="co">#&gt; [170] class_7.3-19                         </span>
<span class="co">#&gt; [171] interactiveDisplayBase_1.31.2        </span>
<span class="co">#&gt; [172] codetools_0.2-18                     </span>
<span class="co">#&gt; [173] utf8_1.2.2                           </span>
<span class="co">#&gt; [174] spatstat.sparse_2.0-0                </span>
<span class="co">#&gt; [175] lattice_0.20-45                      </span>
<span class="co">#&gt; [176] bslib_0.3.1                          </span>
<span class="co">#&gt; [177] ResidualMatrix_1.3.0                 </span>
<span class="co">#&gt; [178] curl_4.3.2                           </span>
<span class="co">#&gt; [179] leiden_0.3.9                         </span>
<span class="co">#&gt; [180] survival_3.2-13                      </span>
<span class="co">#&gt; [181] limma_3.49.5                         </span>
<span class="co">#&gt; [182] rmarkdown_2.11                       </span>
<span class="co">#&gt; [183] desc_1.4.0                           </span>
<span class="co">#&gt; [184] munsell_0.5.0                        </span>
<span class="co">#&gt; [185] e1071_1.7-9                          </span>
<span class="co">#&gt; [186] GetoptLong_1.0.5                     </span>
<span class="co">#&gt; [187] GenomeInfoDbData_1.2.7               </span>
<span class="co">#&gt; [188] iterators_1.0.13                     </span>
<span class="co">#&gt; [189] reshape2_1.4.4                       </span>
<span class="co">#&gt; [190] gtable_0.3.0                         </span>
<span class="co">#&gt; [191] spatstat.core_2.3-0                  </span>
<span class="co">#&gt; [192] Seurat_4.0.5</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-abbas2009deconvolution" class="csl-entry">
Abbas, Alexander R, Kristen Wolslegel, Dhaya Seshasayee, Zora Modrusan, and Hilary F Clark. 2009. <span>“Deconvolution of Blood Microarray Data Identifies Cellular Activation Patterns in Systemic Lupus Erythematosus.”</span> <em>PloS One</em> 4 (7): e6098.
</div>
<div id="ref-aran2019reference" class="csl-entry">
Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, et al. 2019. <span>“Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.”</span> <em>Nature Immunology</em> 20 (2): 163–72.
</div>
<div id="ref-butler2018integrating" class="csl-entry">
Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. 2018. <span>“Integrating Single-Cell Transcriptomic Data Across Different Conditions, Technologies, and Species.”</span> <em>Nature Biotechnology</em> 36 (5): 411–20.
</div>
<div id="ref-Cai2020" class="csl-entry">
Cai, Yi, Youchao Dai, Yejun Wang, Qianqing Yang, Jiubiao Guo, Cailing Wei, Weixin Chen, et al. 2020. <span>“Single-Cell Transcriptomics of Blood Reveals a Natural Killer Cell Subset Depletion in Tuberculosis.”</span> <em><span>EBioMedicine</span></em> 53 (March): 102686. <a href="https://doi.org/10.1016/j.ebiom.2020.102686">https://doi.org/10.1016/j.ebiom.2020.102686</a>.
</div>
<div id="ref-chen2016reads" class="csl-entry">
Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-Ding2020" class="csl-entry">
Ding, Jiarui, Xian Adiconis, Sean K. Simmons, Monika S. Kowalczyk, Cynthia C. Hession, Nemanja D. Marjanovic, Travis K. Hughes, et al. 2020. <span>“Systematic Comparison of Single-Cell and Single-Nucleus <span>RNA</span>-Sequencing Methods.”</span> <em>Nature Biotechnology</em> 38 (6): 737–46. <a href="https://doi.org/10.1038/s41587-020-0465-8">https://doi.org/10.1038/s41587-020-0465-8</a>.
</div>
<div id="ref-Freytag2018" class="csl-entry">
Freytag, Saskia, Luyi Tian, Ingrid Lönnstedt, Milica Ng, and Melanie Bahlo. 2018. <span>“Comparison of Clustering Tools in r for Medium-Sized 10x Genomics Single-Cell <span>RNA</span>-Sequencing Data.”</span> <em>F1000Research</em> 7 (December): 1297. <a href="https://doi.org/10.12688/f1000research.15809.2">https://doi.org/10.12688/f1000research.15809.2</a>.
</div>
<div id="ref-himes2014rna" class="csl-entry">
Himes, Blanca E, Xiaofeng Jiang, Peter Wagner, Ruoxi Hu, Qiyu Wang, Barbara Klanderman, Reid M Whitaker, et al. 2014. <span>“RNA-Seq Transcriptome Profiling Identifies Crispld2 as a Glucocorticoid Responsive Gene That Modulates Cytokine Function in Airway Smooth Muscle Cells.”</span> <em>PloS One</em> 9 (6): e99625.
</div>
<div id="ref-Karagiannis2020" class="csl-entry">
Karagiannis, Tanya T., John P. Cleary, Busra Gok, Andrew J. Henderson, Nicholas G. Martin, Masanao Yajima, Elliot C. Nelson, and Christine S. Cheng. 2020. <span>“Single Cell Transcriptomics Reveals Opioid Usage Evokes Widespread Suppression of Antiviral Gene Program.”</span> <em>Nature Communications</em> 11 (1). <a href="https://doi.org/10.1038/s41467-020-16159-y">https://doi.org/10.1038/s41467-020-16159-y</a>.
</div>
<div id="ref-law2016rna" class="csl-entry">
Law, Charity W, Monther Alhamdoosh, Shian Su, Xueyi Dong, Luyi Tian, Gordon K Smyth, and Matthew E Ritchie. 2016. <span>“RNA-Seq Analysis Is Easy as 1-2-3 with Limma, Glimma and edgeR.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-law2014voom" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. <span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for RNA-Seq Read Counts.”</span> <em>Genome Biology</em> 15 (2): R29.
</div>
<div id="ref-love2014moderated" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15 (12): 550.
</div>
<div id="ref-mccarthy2012differential" class="csl-entry">
McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. <span>“Differential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97.
</div>
<div id="ref-newman2015robust" class="csl-entry">
Newman, Aaron M, Chih Long Liu, Michael R Green, Andrew J Gentles, Weiguo Feng, Yue Xu, Chuong D Hoang, Maximilian Diehn, and Ash A Alizadeh. 2015. <span>“Robust Enumeration of Cell Subsets from Tissue Expression Profiles.”</span> <em>Nature Methods</em> 12 (5): 453–57.
</div>
<div id="ref-racle2017simultaneous" class="csl-entry">
Racle, Julien, Kaat de Jonge, Petra Baumgaertner, Daniel E Speiser, and David Gfeller. 2017. <span>“Simultaneous Enumeration of Cancer and Immune Cell Types from Bulk Tumor Gene Expression Data.”</span> <em>Elife</em> 6: e26476.
</div>
<div id="ref-robinson2010scaling" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.”</span> <em>Genome Biology</em> 11 (3): 1–9.
</div>
<div id="ref-stuart2019comprehensive" class="csl-entry">
Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. <span>“Comprehensive Integration of Single-Cell Data.”</span> <em>Cell</em> 177 (7): 1888–1902.
</div>
<div id="ref-wickham2014tidy" class="csl-entry">
Wickham, Hadley et al. 2014. <span>“Tidy Data.”</span> <em>Journal of Statistical Software</em> 59 (10): 1–23.
</div>
<div id="ref-wickham2019welcome" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the Tidyverse.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686.
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>&lt;maria.doyle at petermac.org&gt;<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>&lt;mangiola.s at wehi.edu.au&gt;<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
